From 1d8b770e83c09cf861c7f32b7c0c3fbfc2fd609a Mon Sep 17 00:00:00 2001
From: timyao <tim.yao@amlogic.com>
Date: Tue, 30 Sep 2014 21:24:08 -0700
Subject: [PATCH 4930/5965] PD #98085 H.264 video decoder: Error frame display
 control moved out of ucode. Avoid entering video PTS discontinious process
 when error frames last more than discontinue threshold.

---
 .../amports/m6/ucode/h264/h264mmc_linux.h     | 160 +++++++++---------
 .../amports/m6tv/ucode/h264/h264mmc_linux.h   | 160 +++++++++---------
 .../amports/m6tvd/ucode/h264/h264mmc_linux.h  | 160 +++++++++---------
 .../amports/m8/ucode/h264/h264mmc_linux.h     | 160 +++++++++---------
 drivers/amlogic/amports/vavs.c                |   1 +
 drivers/amlogic/amports/vh264.c               |  68 +++++---
 drivers/amlogic/amports/vh264_4k2k.c          |   1 +
 drivers/amlogic/amports/vh264mvc.c            |   1 +
 drivers/amlogic/amports/vh265.c               |   1 +
 drivers/amlogic/amports/video.c               |   3 +-
 drivers/amlogic/amports/vmjpeg.c              |   1 +
 drivers/amlogic/amports/vmpeg12.c             |   1 +
 drivers/amlogic/amports/vmpeg4.c              |   3 +
 drivers/amlogic/amports/vreal.c               |   1 +
 drivers/amlogic/amports/vvc1.c                |   3 +
 drivers/amlogic/deinterlace/deinterlace.c     |   1 +
 drivers/amlogic/ppmgr/ppmgr_vpp.c             |   3 +-
 include/linux/amlogic/amports/vframe.h        |   2 +
 18 files changed, 387 insertions(+), 343 deletions(-)

diff --git a/drivers/amlogic/amports/m6/ucode/h264/h264mmc_linux.h b/drivers/amlogic/amports/m6/ucode/h264/h264mmc_linux.h
index d58868f10871..5ffde3db98d7 100755
--- a/drivers/amlogic/amports/m6/ucode/h264/h264mmc_linux.h
+++ b/drivers/amlogic/amports/m6/ucode/h264/h264mmc_linux.h
@@ -5,9 +5,9 @@ const u32 MicroCode[] __attribute__ ((aligned (8))) = {
 	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
 	0xcc00000,0x0400000,0xcc00000,0x0400000,0xcc00000,0x0400000,0xcc00000,0x0400000,
 	0xcc00000,0x0400000,0xc780200,0x6431e07,0xcc00000,0x0400000,0xcc00000,0x0400000,
-	0xcc00000,0x0400000,0x9027207,0xa6140c8,0x0000000,0xc00cec0,0x0000000,0xc780880,
-	0xa614148,0x0000000,0xc003800,0x0000000,0xc780740,0xa614188,0x0000000,0xc006840,
-	0x0000000,0xc780600,0xa6141c8,0x0000000,0xc007000,0x0000000,0xc7804c0,0xa614208,
+	0xcc00000,0x0400000,0x9027207,0xa6140c8,0x0000000,0xc00ce80,0x0000000,0xc780880,
+	0xa614148,0x0000000,0xc003800,0x0000000,0xc780740,0xa614188,0x0000000,0xc006800,
+	0x0000000,0xc780600,0xa6141c8,0x0000000,0xc006fc0,0x0000000,0xc7804c0,0xa614208,
 	0x0000000,0xc000c40,0x0000000,0xc780380,0xa62c288,0x6431f20,0x6433f21,0xe000860,
 	0x0000000,0x0000000,0xf000800,0x6031f20,0xf010840,0xc780100,0x6033f21,0xc780080,
 	0x0000000,0x8031e00,0xcc00000,0x0400000,0x78077c5,0x4002c4e,0x7400c45,0x7807800,
@@ -48,84 +48,84 @@ const u32 MicroCode[] __attribute__ ((aligned (8))) = {
 	0x7400070,0xcc00000,0x0000000,0x680400f,0x5403c4d,0x2031c4f,0x4001c71,0x7000c40,
 	0x6800081,0x9542001,0x7400c40,0x4401c31,0x7000c04,0x649cf30,0xa20c830,0x5808044,
 	0xbe20041,0x7801c01,0xaa10001,0xa216004,0x6800001,0x7c01c01,0x5809040,0xa218041,
-	0x6800001,0x9523001,0x7400c40,0xc781340,0x44012cb,0xc07f400,0x0000000,0x649c001,
-	0xbef7fc1,0x6800001,0x940504d,0x9184c00,0xaa243f0,0x68001c4,0xaa1c030,0x6800004,
-	0xb614170,0x4404130,0xb60c0f0,0x4002130,0x68001c4,0x94a3044,0x7802130,0xaa0c030,
-	0x0000000,0x9801201,0x90c2c00,0xa60c0b0,0x9101c00,0x9801201,0x9521070,0x4401c71,
-	0x7000c70,0x90a2c30,0x9542070,0x78017f0,0x9583070,0x6803230,0x2030370,0x7000c30,
-	0x9610070,0x6800030,0x2030370,0x7000c04,0x5402c0b,0x20302f0,0x6836eb1,0x2031c31,
-	0xcb80031,0x0000000,0x609ca04,0x0000000,0x0000000,0xc7806c0,0x609c101,0x649ca30,
-	0x9610c04,0x609ca30,0xc780580,0x609c201,0x609cb04,0x0000000,0x0000000,0xc780440,
-	0x609c301,0x649cb30,0x9610c04,0x609cb30,0xc780300,0x609c401,0x609cc04,0x0000000,
-	0x0000000,0xc7801c0,0x609c501,0x649cc30,0x9610c04,0x609cc30,0xc780080,0x609c601,
-	0x649cf01,0xbe09001,0xbe100c0,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x0000000,
-	0x680000b,0x7803d0d,0xa21880d,0x780318c,0x7801b8d,0xb00334c,0x0000000,0x7801b8c,
-	0x78071cd,0x649cf00,0xbe0d000,0xb01230d,0xc7800c0,0x680004c,0xb80f30d,0x0000000,
-	0xc07d440,0x0000000,0xa61bfcd,0x0000000,0xc079e00,0x0000000,0xc7801c0,0x0000000,
-	0xc07dec0,0x0000000,0x40012cb,0xaa0818b,0xc7ff940,0xaa1400b,0x6800080,0x950800b,
-	0x609c000,0x609ce00,0xcc00000,0x0000000,0x680000b,0x78071cd,0xaa4800d,0x0000000,
-	0xc07ce40,0x0000000,0xa61bfcd,0x0000000,0xc079800,0x0000000,0xc780280,0x0000000,
-	0x7807180,0xa807340,0x0000000,0xc07d800,0x0000000,0x40012cb,0xaa0818b,0xc7ffb80,
-	0xaa1c00b,0x6800080,0x950800b,0x609c000,0x609ce00,0xc7ffa00,0x680000b,0xcc00000,
-	0x0000000,0x680400f,0x780704e,0x6800008,0x6800007,0x7807186,0x78072c0,0xa225800,
-	0x7c07087,0x5403c46,0x2031c4f,0x7000c70,0x9005c30,0x6800041,0x3401c01,0x7c07081,
-	0x5403c48,0x2031c4f,0x4001c71,0x7000c70,0xa23c0f0,0xa406188,0x6800000,0x9444c00,
-	0x7400c70,0xc780280,0x6800047,0x6800000,0x9406c00,0x7400c70,0x6810000,0xbc04030,
-	0x78071c0,0x4401000,0x7c071c0,0x4001208,0xb8ec388,0xcc00000,0x7c07207,0x680400f,
-	0x780704e,0x6800008,0x6800007,0x7807186,0x5403c48,0x2031c4f,0x4001c71,0x7000c70,
-	0x6810000,0xa007030,0xa805188,0x0000000,0x9c302b0,0xc7800c0,0x7400c70,0x6800047,
-	0x4001208,0xb8f3388,0xcc00000,0x7c071c7,0x400138e,0x780734d,0x240e38d,0x7807300,
-	0xb610080,0x0000000,0x200e34e,0x400138e,0x6800008,0x7807109,0x680a00a,0xa403209,
-	0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,
-	0x9610001,0xa42a380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,
-	0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa18080,0x6800000,0xc780180,0x94421c0,
-	0xc780100,0x9c07087,0xc780080,0x9c070c7,0xbe4c307,0x6800000,0x94021c0,0x7807200,
-	0x4401000,0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,
-	0x3000040,0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0x649cf00,
-	0xbe14080,0x7400187,0xbe0d000,0x4001208,0xb8ce248,0xcc00000,0x0000000,0x7807108,
-	0x7807149,0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,
-	0x7000c40,0x4001c71,0x7000c41,0x9610001,0xa42a380,0x7000105,0x6804031,0x5403005,
-	0x2006031,0x4001186,0x7000187,0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa1c080,
-	0x6800000,0xc780200,0x94441c0,0x9c07087,0xc780140,0x9c07107,0x9c070c7,0xc780080,
-	0x9c07147,0xbe4c307,0x6800000,0x94021c0,0x7807200,0x4401000,0x7c07200,0x4401046,
-	0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,0x7c07080,0x6810000,
-	0xbc04007,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x7400187,0x4001208,0xb8ce248,
-	0xcc00000,0x0000000,0x680000d,0x7807300,0xb60c080,0x0000000,0x680004d,0x7807108,
-	0x7807149,0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,
-	0x7000c40,0x4001c71,0x7000c41,0x9610001,0x3800340,0xa41c380,0x7000105,0x6804031,
-	0x5403005,0x2006031,0x4001186,0x7000187,0xa24c307,0x6800000,0x94061c0,0x7807200,
-	0x4401000,0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,
-	0x3000040,0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0xcc00000,
-	0x7400187,0x4001208,0xb8db248,0xcc00000,0x0000000,0x400134d,0x780734c,0x240d34c,
-	0x7807300,0xb610080,0x0000000,0x200d30d,0x400134d,0x6800008,0x7807109,0x680a00a,
-	0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,
-	0x7000c41,0x9610001,0xa417340,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,
-	0x7000187,0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa18080,0x68000c0,0xc780180,
-	0x94821c0,0xc780100,0x9807107,0xc780080,0x9807147,0x4006c46,0x7400c4e,0xcc00000,
-	0x7400187,0x4001208,0xb8e1248,0xcc00000,0x0000000,0x440138e,0x680000d,0x7807300,
-	0xb60c080,0x0000000,0x680004d,0x7807108,0x7807149,0x680a00a,0xa403209,0xcc00000,
+	0x6800001,0x9523001,0x7400c40,0xc781300,0x44012cb,0xc07f400,0x0000000,0x649c001,
+	0xbef7fc1,0x9101044,0x5409041,0x940504d,0x9184c00,0xaa243f0,0x68001c4,0xaa1c030,
+	0x6800004,0xb614170,0x4404130,0xb60c0f0,0x4002130,0x68001c4,0x94a3044,0x7802130,
+	0xaa0c030,0x0000000,0x9801201,0xa60c0b0,0x90c2c00,0x9801201,0x4401c71,0x7000c70,
+	0x90a2c30,0x9542070,0x78017f0,0x9583070,0x6803230,0x2030370,0x7000c30,0x9610070,
+	0x6800030,0x2030370,0x7000c04,0x5402c0b,0x20302f0,0x6836e71,0x2031c31,0xcb80031,
+	0x0000000,0x609ca04,0x0000000,0x0000000,0xc7806c0,0x609c101,0x649ca30,0x9610c04,
+	0x609ca30,0xc780580,0x609c201,0x609cb04,0x0000000,0x0000000,0xc780440,0x609c301,
+	0x649cb30,0x9610c04,0x609cb30,0xc780300,0x609c401,0x609cc04,0x0000000,0x0000000,
+	0xc7801c0,0x609c501,0x649cc30,0x9610c04,0x609cc30,0xc780080,0x609c601,0x649cf01,
+	0xbe09001,0xbe100c0,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x0000000,0x680000b,
+	0x7803d0d,0xa21880d,0x780318c,0x7801b8d,0xb00334c,0x0000000,0x7801b8c,0x78071cd,
+	0x649cf00,0xbe0d000,0xb01230d,0xc7800c0,0x680004c,0xb80f30d,0x0000000,0xc07d480,
+	0x0000000,0xa61bfcd,0x0000000,0xc079e40,0x0000000,0xc7801c0,0x0000000,0xc07df00,
+	0x0000000,0x40012cb,0xaa0818b,0xc7ff940,0xaa1400b,0x6800080,0x950800b,0x609c000,
+	0x609ce00,0xcc00000,0x0000000,0x680000b,0x78071cd,0xaa4800d,0x0000000,0xc07ce80,
+	0x0000000,0xa61bfcd,0x0000000,0xc079840,0x0000000,0xc780280,0x0000000,0x7807180,
+	0xa807340,0x0000000,0xc07d840,0x0000000,0x40012cb,0xaa0818b,0xc7ffb80,0xaa1c00b,
+	0x6800080,0x950800b,0x609c000,0x609ce00,0xc7ffa00,0x680000b,0xcc00000,0x0000000,
+	0x680400f,0x780704e,0x6800008,0x6800007,0x7807186,0x78072c0,0xa225800,0x7c07087,
+	0x5403c46,0x2031c4f,0x7000c70,0x9005c30,0x6800041,0x3401c01,0x7c07081,0x5403c48,
+	0x2031c4f,0x4001c71,0x7000c70,0xa23c0f0,0xa406188,0x6800000,0x9444c00,0x7400c70,
+	0xc780280,0x6800047,0x6800000,0x9406c00,0x7400c70,0x6810000,0xbc04030,0x78071c0,
+	0x4401000,0x7c071c0,0x4001208,0xb8ec388,0xcc00000,0x7c07207,0x680400f,0x780704e,
+	0x6800008,0x6800007,0x7807186,0x5403c48,0x2031c4f,0x4001c71,0x7000c70,0x6810000,
+	0xa007030,0xa805188,0x0000000,0x9c302b0,0xc7800c0,0x7400c70,0x6800047,0x4001208,
+	0xb8f3388,0xcc00000,0x7c071c7,0x400138e,0x780734d,0x240e38d,0x7807300,0xb610080,
+	0x0000000,0x200e34e,0x400138e,0x6800008,0x7807109,0x680a00a,0xa403209,0xcc00000,
 	0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,0x9610001,
-	0x3800340,0xb01b380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,
-	0xa250307,0x6800000,0x94061c0,0x7807200,0x4401000,0x7c07200,0x4401046,0x7000040,
+	0xa42a380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,0x4001c44,
+	0x7000c70,0x9002030,0xaa14040,0xaa18080,0x6800000,0xc780180,0x94421c0,0xc780100,
+	0x9c07087,0xc780080,0x9c070c7,0xbe4c307,0x6800000,0x94021c0,0x7807200,0x4401000,
+	0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,
+	0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0x649cf00,0xbe14080,
+	0x7400187,0xbe0d000,0x4001208,0xb8ce248,0xcc00000,0x0000000,0x7807108,0x7807149,
+	0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,
+	0x4001c71,0x7000c41,0x9610001,0xa42a380,0x7000105,0x6804031,0x5403005,0x2006031,
+	0x4001186,0x7000187,0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa1c080,0x6800000,
+	0xc780200,0x94441c0,0x9c07087,0xc780140,0x9c07107,0x9c070c7,0xc780080,0x9c07147,
+	0xbe4c307,0x6800000,0x94021c0,0x7807200,0x4401000,0x7c07200,0x4401046,0x7000040,
 	0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,0x7c07080,0x6810000,0xbc04007,
-	0x78071c0,0x4401000,0x7c071c0,0x7400187,0x4001208,0xb8dc248,0xcc00000,0x0000000,
-	0x6800010,0x6800011,0x78072f1,0x9005031,0xa638140,0x0000000,0xc07aac0,0x0000000,
-	0x780e100,0xbe18080,0x9001400,0xc07a240,0x0000000,0xc781900,0x0000000,0xc07b200,
-	0x0000000,0xc781800,0x6800000,0x90a2071,0x780e130,0xa85c001,0xa85b030,0x680e10f,
-	0x70003c0,0xaa2c040,0xaa40080,0xaa540c0,0xaa78100,0x0000000,0x4605040,0xc500840,
-	0x0000000,0x4606040,0xc5011c0,0xc781380,0x4001c4f,0x7000c4e,0xc07b180,0x0000000,
-	0xc7ffc00,0x40023cf,0x4001c4f,0x7000c4e,0xc07c0c0,0x0000000,0xc7ffa80,0x40023cf,
-	0x4002c4f,0x7000c4e,0xc07ce00,0x0000000,0x4001c4f,0x7000c4d,0xc07d9c0,0x0000000,
-	0xc7ff800,0x40033cf,0x4001c4f,0x7000c4e,0xc07e440,0x0000000,0xc7ff680,0x40023cf,
-	0xc079c40,0x0000000,0xc079480,0x0000000,0x6800007,0x7807331,0xb250071,0x7807770,
-	0x7807780,0x9610c00,0x7807841,0x7807880,0x9610040,0x2401c01,0x7c07841,0x5810001,
-	0x7c07880,0x78077c1,0x7807800,0x9610040,0x2401c01,0x7c077c1,0x5810001,0x7c07800,
-	0xc7801c0,0x7c07441,0x7c07447,0x7c077c7,0x7c07807,0x7c07847,0x7c07887,0x7c07347,
-	0x7c07487,0x7c07647,0x7c07687,0x7c076c7,0x7c07707,0x7c07747,0x7c07787,0xc7fec40,
-	0x40013cf,0x4001c4f,0x7000c4e,0xc07bfc0,0x0000000,0x6800050,0x400044e,0xc7fea40,
-	0x40023cf,0x680400f,0x780718e,0x540338e,0x200e38f,0x4001c4e,0x7000c40,0xa2140c0,
-	0xc072480,0x7c031d0,0xcc00000,0x0000000,0x78072f1,0x9005031,0x90a2071,0xaa10140,
-	0xaa0c001,0xc074a40,0x0000000,0x649cf00,0xa219000,0xc072140,0x7c031d0,0xc077a40,
-	0x0000000,0xc780140,0xc077980,0x0000000,0xc071f80,0x7c031d0,0xcc00000,0x0000000,
+	0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x7400187,0x4001208,0xb8ce248,0xcc00000,
+	0x0000000,0x680000d,0x7807300,0xb60c080,0x0000000,0x680004d,0x7807108,0x7807149,
+	0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,
+	0x4001c71,0x7000c41,0x9610001,0x3800340,0xa41c380,0x7000105,0x6804031,0x5403005,
+	0x2006031,0x4001186,0x7000187,0xa24c307,0x6800000,0x94061c0,0x7807200,0x4401000,
+	0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,
+	0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x7400187,
+	0x4001208,0xb8db248,0xcc00000,0x0000000,0x400134d,0x780734c,0x240d34c,0x7807300,
+	0xb610080,0x0000000,0x200d30d,0x400134d,0x6800008,0x7807109,0x680a00a,0xa403209,
+	0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,
+	0x9610001,0xa417340,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,
+	0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa18080,0x68000c0,0xc780180,0x94821c0,
+	0xc780100,0x9807107,0xc780080,0x9807147,0x4006c46,0x7400c4e,0xcc00000,0x7400187,
+	0x4001208,0xb8e1248,0xcc00000,0x0000000,0x440138e,0x680000d,0x7807300,0xb60c080,
+	0x0000000,0x680004d,0x7807108,0x7807149,0x680a00a,0xa403209,0xcc00000,0x0000000,
+	0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,0x9610001,0x3800340,
+	0xb01b380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,0xa250307,
+	0x6800000,0x94061c0,0x7807200,0x4401000,0x7c07200,0x4401046,0x7000040,0x9005000,
+	0x6800041,0x3401001,0x7807080,0x3000040,0x7c07080,0x6810000,0xbc04007,0x78071c0,
+	0x4401000,0x7c071c0,0x7400187,0x4001208,0xb8dc248,0xcc00000,0x0000000,0x6800010,
+	0x6800011,0x78072f1,0x9005031,0xa638140,0x0000000,0xc07aac0,0x0000000,0x780e100,
+	0xbe18080,0x9001400,0xc07a240,0x0000000,0xc781900,0x0000000,0xc07b200,0x0000000,
+	0xc781800,0x6800000,0x90a2071,0x780e130,0xa85c001,0xa85b030,0x680e10f,0x70003c0,
+	0xaa2c040,0xaa40080,0xaa540c0,0xaa78100,0x0000000,0x4605040,0xc500840,0x0000000,
+	0x4606040,0xc5011c0,0xc781380,0x4001c4f,0x7000c4e,0xc07b180,0x0000000,0xc7ffc00,
+	0x40023cf,0x4001c4f,0x7000c4e,0xc07c0c0,0x0000000,0xc7ffa80,0x40023cf,0x4002c4f,
+	0x7000c4e,0xc07ce00,0x0000000,0x4001c4f,0x7000c4d,0xc07d9c0,0x0000000,0xc7ff800,
+	0x40033cf,0x4001c4f,0x7000c4e,0xc07e440,0x0000000,0xc7ff680,0x40023cf,0xc079c40,
+	0x0000000,0xc079480,0x0000000,0x6800007,0x7807331,0xb250071,0x7807770,0x7807780,
+	0x9610c00,0x7807841,0x7807880,0x9610040,0x2401c01,0x7c07841,0x5810001,0x7c07880,
+	0x78077c1,0x7807800,0x9610040,0x2401c01,0x7c077c1,0x5810001,0x7c07800,0xc7801c0,
+	0x7c07441,0x7c07447,0x7c077c7,0x7c07807,0x7c07847,0x7c07887,0x7c07347,0x7c07487,
+	0x7c07647,0x7c07687,0x7c076c7,0x7c07707,0x7c07747,0x7c07787,0xc7fec40,0x40013cf,
+	0x4001c4f,0x7000c4e,0xc07bfc0,0x0000000,0x6800050,0x400044e,0xc7fea40,0x40023cf,
+	0x680400f,0x780718e,0x540338e,0x200e38f,0x4001c4e,0x7000c40,0xa2140c0,0xc0724c0,
+	0x7c031d0,0xcc00000,0x0000000,0x78072f1,0x9005031,0x90a2071,0xaa10140,0xaa0c001,
+	0xc074a80,0x0000000,0x649cf00,0xa219000,0xc072180,0x7c031d0,0xc077a40,0x0000000,
+	0xc780140,0xc077980,0x0000000,0xc071fc0,0x7c031d0,0xcc00000,0x0000000,0x0000000,
 	0x0000000
 };
diff --git a/drivers/amlogic/amports/m6tv/ucode/h264/h264mmc_linux.h b/drivers/amlogic/amports/m6tv/ucode/h264/h264mmc_linux.h
index d58868f10871..5ffde3db98d7 100755
--- a/drivers/amlogic/amports/m6tv/ucode/h264/h264mmc_linux.h
+++ b/drivers/amlogic/amports/m6tv/ucode/h264/h264mmc_linux.h
@@ -5,9 +5,9 @@ const u32 MicroCode[] __attribute__ ((aligned (8))) = {
 	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
 	0xcc00000,0x0400000,0xcc00000,0x0400000,0xcc00000,0x0400000,0xcc00000,0x0400000,
 	0xcc00000,0x0400000,0xc780200,0x6431e07,0xcc00000,0x0400000,0xcc00000,0x0400000,
-	0xcc00000,0x0400000,0x9027207,0xa6140c8,0x0000000,0xc00cec0,0x0000000,0xc780880,
-	0xa614148,0x0000000,0xc003800,0x0000000,0xc780740,0xa614188,0x0000000,0xc006840,
-	0x0000000,0xc780600,0xa6141c8,0x0000000,0xc007000,0x0000000,0xc7804c0,0xa614208,
+	0xcc00000,0x0400000,0x9027207,0xa6140c8,0x0000000,0xc00ce80,0x0000000,0xc780880,
+	0xa614148,0x0000000,0xc003800,0x0000000,0xc780740,0xa614188,0x0000000,0xc006800,
+	0x0000000,0xc780600,0xa6141c8,0x0000000,0xc006fc0,0x0000000,0xc7804c0,0xa614208,
 	0x0000000,0xc000c40,0x0000000,0xc780380,0xa62c288,0x6431f20,0x6433f21,0xe000860,
 	0x0000000,0x0000000,0xf000800,0x6031f20,0xf010840,0xc780100,0x6033f21,0xc780080,
 	0x0000000,0x8031e00,0xcc00000,0x0400000,0x78077c5,0x4002c4e,0x7400c45,0x7807800,
@@ -48,84 +48,84 @@ const u32 MicroCode[] __attribute__ ((aligned (8))) = {
 	0x7400070,0xcc00000,0x0000000,0x680400f,0x5403c4d,0x2031c4f,0x4001c71,0x7000c40,
 	0x6800081,0x9542001,0x7400c40,0x4401c31,0x7000c04,0x649cf30,0xa20c830,0x5808044,
 	0xbe20041,0x7801c01,0xaa10001,0xa216004,0x6800001,0x7c01c01,0x5809040,0xa218041,
-	0x6800001,0x9523001,0x7400c40,0xc781340,0x44012cb,0xc07f400,0x0000000,0x649c001,
-	0xbef7fc1,0x6800001,0x940504d,0x9184c00,0xaa243f0,0x68001c4,0xaa1c030,0x6800004,
-	0xb614170,0x4404130,0xb60c0f0,0x4002130,0x68001c4,0x94a3044,0x7802130,0xaa0c030,
-	0x0000000,0x9801201,0x90c2c00,0xa60c0b0,0x9101c00,0x9801201,0x9521070,0x4401c71,
-	0x7000c70,0x90a2c30,0x9542070,0x78017f0,0x9583070,0x6803230,0x2030370,0x7000c30,
-	0x9610070,0x6800030,0x2030370,0x7000c04,0x5402c0b,0x20302f0,0x6836eb1,0x2031c31,
-	0xcb80031,0x0000000,0x609ca04,0x0000000,0x0000000,0xc7806c0,0x609c101,0x649ca30,
-	0x9610c04,0x609ca30,0xc780580,0x609c201,0x609cb04,0x0000000,0x0000000,0xc780440,
-	0x609c301,0x649cb30,0x9610c04,0x609cb30,0xc780300,0x609c401,0x609cc04,0x0000000,
-	0x0000000,0xc7801c0,0x609c501,0x649cc30,0x9610c04,0x609cc30,0xc780080,0x609c601,
-	0x649cf01,0xbe09001,0xbe100c0,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x0000000,
-	0x680000b,0x7803d0d,0xa21880d,0x780318c,0x7801b8d,0xb00334c,0x0000000,0x7801b8c,
-	0x78071cd,0x649cf00,0xbe0d000,0xb01230d,0xc7800c0,0x680004c,0xb80f30d,0x0000000,
-	0xc07d440,0x0000000,0xa61bfcd,0x0000000,0xc079e00,0x0000000,0xc7801c0,0x0000000,
-	0xc07dec0,0x0000000,0x40012cb,0xaa0818b,0xc7ff940,0xaa1400b,0x6800080,0x950800b,
-	0x609c000,0x609ce00,0xcc00000,0x0000000,0x680000b,0x78071cd,0xaa4800d,0x0000000,
-	0xc07ce40,0x0000000,0xa61bfcd,0x0000000,0xc079800,0x0000000,0xc780280,0x0000000,
-	0x7807180,0xa807340,0x0000000,0xc07d800,0x0000000,0x40012cb,0xaa0818b,0xc7ffb80,
-	0xaa1c00b,0x6800080,0x950800b,0x609c000,0x609ce00,0xc7ffa00,0x680000b,0xcc00000,
-	0x0000000,0x680400f,0x780704e,0x6800008,0x6800007,0x7807186,0x78072c0,0xa225800,
-	0x7c07087,0x5403c46,0x2031c4f,0x7000c70,0x9005c30,0x6800041,0x3401c01,0x7c07081,
-	0x5403c48,0x2031c4f,0x4001c71,0x7000c70,0xa23c0f0,0xa406188,0x6800000,0x9444c00,
-	0x7400c70,0xc780280,0x6800047,0x6800000,0x9406c00,0x7400c70,0x6810000,0xbc04030,
-	0x78071c0,0x4401000,0x7c071c0,0x4001208,0xb8ec388,0xcc00000,0x7c07207,0x680400f,
-	0x780704e,0x6800008,0x6800007,0x7807186,0x5403c48,0x2031c4f,0x4001c71,0x7000c70,
-	0x6810000,0xa007030,0xa805188,0x0000000,0x9c302b0,0xc7800c0,0x7400c70,0x6800047,
-	0x4001208,0xb8f3388,0xcc00000,0x7c071c7,0x400138e,0x780734d,0x240e38d,0x7807300,
-	0xb610080,0x0000000,0x200e34e,0x400138e,0x6800008,0x7807109,0x680a00a,0xa403209,
-	0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,
-	0x9610001,0xa42a380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,
-	0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa18080,0x6800000,0xc780180,0x94421c0,
-	0xc780100,0x9c07087,0xc780080,0x9c070c7,0xbe4c307,0x6800000,0x94021c0,0x7807200,
-	0x4401000,0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,
-	0x3000040,0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0x649cf00,
-	0xbe14080,0x7400187,0xbe0d000,0x4001208,0xb8ce248,0xcc00000,0x0000000,0x7807108,
-	0x7807149,0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,
-	0x7000c40,0x4001c71,0x7000c41,0x9610001,0xa42a380,0x7000105,0x6804031,0x5403005,
-	0x2006031,0x4001186,0x7000187,0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa1c080,
-	0x6800000,0xc780200,0x94441c0,0x9c07087,0xc780140,0x9c07107,0x9c070c7,0xc780080,
-	0x9c07147,0xbe4c307,0x6800000,0x94021c0,0x7807200,0x4401000,0x7c07200,0x4401046,
-	0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,0x7c07080,0x6810000,
-	0xbc04007,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x7400187,0x4001208,0xb8ce248,
-	0xcc00000,0x0000000,0x680000d,0x7807300,0xb60c080,0x0000000,0x680004d,0x7807108,
-	0x7807149,0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,
-	0x7000c40,0x4001c71,0x7000c41,0x9610001,0x3800340,0xa41c380,0x7000105,0x6804031,
-	0x5403005,0x2006031,0x4001186,0x7000187,0xa24c307,0x6800000,0x94061c0,0x7807200,
-	0x4401000,0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,
-	0x3000040,0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0xcc00000,
-	0x7400187,0x4001208,0xb8db248,0xcc00000,0x0000000,0x400134d,0x780734c,0x240d34c,
-	0x7807300,0xb610080,0x0000000,0x200d30d,0x400134d,0x6800008,0x7807109,0x680a00a,
-	0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,
-	0x7000c41,0x9610001,0xa417340,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,
-	0x7000187,0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa18080,0x68000c0,0xc780180,
-	0x94821c0,0xc780100,0x9807107,0xc780080,0x9807147,0x4006c46,0x7400c4e,0xcc00000,
-	0x7400187,0x4001208,0xb8e1248,0xcc00000,0x0000000,0x440138e,0x680000d,0x7807300,
-	0xb60c080,0x0000000,0x680004d,0x7807108,0x7807149,0x680a00a,0xa403209,0xcc00000,
+	0x6800001,0x9523001,0x7400c40,0xc781300,0x44012cb,0xc07f400,0x0000000,0x649c001,
+	0xbef7fc1,0x9101044,0x5409041,0x940504d,0x9184c00,0xaa243f0,0x68001c4,0xaa1c030,
+	0x6800004,0xb614170,0x4404130,0xb60c0f0,0x4002130,0x68001c4,0x94a3044,0x7802130,
+	0xaa0c030,0x0000000,0x9801201,0xa60c0b0,0x90c2c00,0x9801201,0x4401c71,0x7000c70,
+	0x90a2c30,0x9542070,0x78017f0,0x9583070,0x6803230,0x2030370,0x7000c30,0x9610070,
+	0x6800030,0x2030370,0x7000c04,0x5402c0b,0x20302f0,0x6836e71,0x2031c31,0xcb80031,
+	0x0000000,0x609ca04,0x0000000,0x0000000,0xc7806c0,0x609c101,0x649ca30,0x9610c04,
+	0x609ca30,0xc780580,0x609c201,0x609cb04,0x0000000,0x0000000,0xc780440,0x609c301,
+	0x649cb30,0x9610c04,0x609cb30,0xc780300,0x609c401,0x609cc04,0x0000000,0x0000000,
+	0xc7801c0,0x609c501,0x649cc30,0x9610c04,0x609cc30,0xc780080,0x609c601,0x649cf01,
+	0xbe09001,0xbe100c0,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x0000000,0x680000b,
+	0x7803d0d,0xa21880d,0x780318c,0x7801b8d,0xb00334c,0x0000000,0x7801b8c,0x78071cd,
+	0x649cf00,0xbe0d000,0xb01230d,0xc7800c0,0x680004c,0xb80f30d,0x0000000,0xc07d480,
+	0x0000000,0xa61bfcd,0x0000000,0xc079e40,0x0000000,0xc7801c0,0x0000000,0xc07df00,
+	0x0000000,0x40012cb,0xaa0818b,0xc7ff940,0xaa1400b,0x6800080,0x950800b,0x609c000,
+	0x609ce00,0xcc00000,0x0000000,0x680000b,0x78071cd,0xaa4800d,0x0000000,0xc07ce80,
+	0x0000000,0xa61bfcd,0x0000000,0xc079840,0x0000000,0xc780280,0x0000000,0x7807180,
+	0xa807340,0x0000000,0xc07d840,0x0000000,0x40012cb,0xaa0818b,0xc7ffb80,0xaa1c00b,
+	0x6800080,0x950800b,0x609c000,0x609ce00,0xc7ffa00,0x680000b,0xcc00000,0x0000000,
+	0x680400f,0x780704e,0x6800008,0x6800007,0x7807186,0x78072c0,0xa225800,0x7c07087,
+	0x5403c46,0x2031c4f,0x7000c70,0x9005c30,0x6800041,0x3401c01,0x7c07081,0x5403c48,
+	0x2031c4f,0x4001c71,0x7000c70,0xa23c0f0,0xa406188,0x6800000,0x9444c00,0x7400c70,
+	0xc780280,0x6800047,0x6800000,0x9406c00,0x7400c70,0x6810000,0xbc04030,0x78071c0,
+	0x4401000,0x7c071c0,0x4001208,0xb8ec388,0xcc00000,0x7c07207,0x680400f,0x780704e,
+	0x6800008,0x6800007,0x7807186,0x5403c48,0x2031c4f,0x4001c71,0x7000c70,0x6810000,
+	0xa007030,0xa805188,0x0000000,0x9c302b0,0xc7800c0,0x7400c70,0x6800047,0x4001208,
+	0xb8f3388,0xcc00000,0x7c071c7,0x400138e,0x780734d,0x240e38d,0x7807300,0xb610080,
+	0x0000000,0x200e34e,0x400138e,0x6800008,0x7807109,0x680a00a,0xa403209,0xcc00000,
 	0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,0x9610001,
-	0x3800340,0xb01b380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,
-	0xa250307,0x6800000,0x94061c0,0x7807200,0x4401000,0x7c07200,0x4401046,0x7000040,
+	0xa42a380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,0x4001c44,
+	0x7000c70,0x9002030,0xaa14040,0xaa18080,0x6800000,0xc780180,0x94421c0,0xc780100,
+	0x9c07087,0xc780080,0x9c070c7,0xbe4c307,0x6800000,0x94021c0,0x7807200,0x4401000,
+	0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,
+	0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0x649cf00,0xbe14080,
+	0x7400187,0xbe0d000,0x4001208,0xb8ce248,0xcc00000,0x0000000,0x7807108,0x7807149,
+	0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,
+	0x4001c71,0x7000c41,0x9610001,0xa42a380,0x7000105,0x6804031,0x5403005,0x2006031,
+	0x4001186,0x7000187,0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa1c080,0x6800000,
+	0xc780200,0x94441c0,0x9c07087,0xc780140,0x9c07107,0x9c070c7,0xc780080,0x9c07147,
+	0xbe4c307,0x6800000,0x94021c0,0x7807200,0x4401000,0x7c07200,0x4401046,0x7000040,
 	0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,0x7c07080,0x6810000,0xbc04007,
-	0x78071c0,0x4401000,0x7c071c0,0x7400187,0x4001208,0xb8dc248,0xcc00000,0x0000000,
-	0x6800010,0x6800011,0x78072f1,0x9005031,0xa638140,0x0000000,0xc07aac0,0x0000000,
-	0x780e100,0xbe18080,0x9001400,0xc07a240,0x0000000,0xc781900,0x0000000,0xc07b200,
-	0x0000000,0xc781800,0x6800000,0x90a2071,0x780e130,0xa85c001,0xa85b030,0x680e10f,
-	0x70003c0,0xaa2c040,0xaa40080,0xaa540c0,0xaa78100,0x0000000,0x4605040,0xc500840,
-	0x0000000,0x4606040,0xc5011c0,0xc781380,0x4001c4f,0x7000c4e,0xc07b180,0x0000000,
-	0xc7ffc00,0x40023cf,0x4001c4f,0x7000c4e,0xc07c0c0,0x0000000,0xc7ffa80,0x40023cf,
-	0x4002c4f,0x7000c4e,0xc07ce00,0x0000000,0x4001c4f,0x7000c4d,0xc07d9c0,0x0000000,
-	0xc7ff800,0x40033cf,0x4001c4f,0x7000c4e,0xc07e440,0x0000000,0xc7ff680,0x40023cf,
-	0xc079c40,0x0000000,0xc079480,0x0000000,0x6800007,0x7807331,0xb250071,0x7807770,
-	0x7807780,0x9610c00,0x7807841,0x7807880,0x9610040,0x2401c01,0x7c07841,0x5810001,
-	0x7c07880,0x78077c1,0x7807800,0x9610040,0x2401c01,0x7c077c1,0x5810001,0x7c07800,
-	0xc7801c0,0x7c07441,0x7c07447,0x7c077c7,0x7c07807,0x7c07847,0x7c07887,0x7c07347,
-	0x7c07487,0x7c07647,0x7c07687,0x7c076c7,0x7c07707,0x7c07747,0x7c07787,0xc7fec40,
-	0x40013cf,0x4001c4f,0x7000c4e,0xc07bfc0,0x0000000,0x6800050,0x400044e,0xc7fea40,
-	0x40023cf,0x680400f,0x780718e,0x540338e,0x200e38f,0x4001c4e,0x7000c40,0xa2140c0,
-	0xc072480,0x7c031d0,0xcc00000,0x0000000,0x78072f1,0x9005031,0x90a2071,0xaa10140,
-	0xaa0c001,0xc074a40,0x0000000,0x649cf00,0xa219000,0xc072140,0x7c031d0,0xc077a40,
-	0x0000000,0xc780140,0xc077980,0x0000000,0xc071f80,0x7c031d0,0xcc00000,0x0000000,
+	0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x7400187,0x4001208,0xb8ce248,0xcc00000,
+	0x0000000,0x680000d,0x7807300,0xb60c080,0x0000000,0x680004d,0x7807108,0x7807149,
+	0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,
+	0x4001c71,0x7000c41,0x9610001,0x3800340,0xa41c380,0x7000105,0x6804031,0x5403005,
+	0x2006031,0x4001186,0x7000187,0xa24c307,0x6800000,0x94061c0,0x7807200,0x4401000,
+	0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,
+	0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x7400187,
+	0x4001208,0xb8db248,0xcc00000,0x0000000,0x400134d,0x780734c,0x240d34c,0x7807300,
+	0xb610080,0x0000000,0x200d30d,0x400134d,0x6800008,0x7807109,0x680a00a,0xa403209,
+	0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,
+	0x9610001,0xa417340,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,
+	0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa18080,0x68000c0,0xc780180,0x94821c0,
+	0xc780100,0x9807107,0xc780080,0x9807147,0x4006c46,0x7400c4e,0xcc00000,0x7400187,
+	0x4001208,0xb8e1248,0xcc00000,0x0000000,0x440138e,0x680000d,0x7807300,0xb60c080,
+	0x0000000,0x680004d,0x7807108,0x7807149,0x680a00a,0xa403209,0xcc00000,0x0000000,
+	0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,0x9610001,0x3800340,
+	0xb01b380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,0xa250307,
+	0x6800000,0x94061c0,0x7807200,0x4401000,0x7c07200,0x4401046,0x7000040,0x9005000,
+	0x6800041,0x3401001,0x7807080,0x3000040,0x7c07080,0x6810000,0xbc04007,0x78071c0,
+	0x4401000,0x7c071c0,0x7400187,0x4001208,0xb8dc248,0xcc00000,0x0000000,0x6800010,
+	0x6800011,0x78072f1,0x9005031,0xa638140,0x0000000,0xc07aac0,0x0000000,0x780e100,
+	0xbe18080,0x9001400,0xc07a240,0x0000000,0xc781900,0x0000000,0xc07b200,0x0000000,
+	0xc781800,0x6800000,0x90a2071,0x780e130,0xa85c001,0xa85b030,0x680e10f,0x70003c0,
+	0xaa2c040,0xaa40080,0xaa540c0,0xaa78100,0x0000000,0x4605040,0xc500840,0x0000000,
+	0x4606040,0xc5011c0,0xc781380,0x4001c4f,0x7000c4e,0xc07b180,0x0000000,0xc7ffc00,
+	0x40023cf,0x4001c4f,0x7000c4e,0xc07c0c0,0x0000000,0xc7ffa80,0x40023cf,0x4002c4f,
+	0x7000c4e,0xc07ce00,0x0000000,0x4001c4f,0x7000c4d,0xc07d9c0,0x0000000,0xc7ff800,
+	0x40033cf,0x4001c4f,0x7000c4e,0xc07e440,0x0000000,0xc7ff680,0x40023cf,0xc079c40,
+	0x0000000,0xc079480,0x0000000,0x6800007,0x7807331,0xb250071,0x7807770,0x7807780,
+	0x9610c00,0x7807841,0x7807880,0x9610040,0x2401c01,0x7c07841,0x5810001,0x7c07880,
+	0x78077c1,0x7807800,0x9610040,0x2401c01,0x7c077c1,0x5810001,0x7c07800,0xc7801c0,
+	0x7c07441,0x7c07447,0x7c077c7,0x7c07807,0x7c07847,0x7c07887,0x7c07347,0x7c07487,
+	0x7c07647,0x7c07687,0x7c076c7,0x7c07707,0x7c07747,0x7c07787,0xc7fec40,0x40013cf,
+	0x4001c4f,0x7000c4e,0xc07bfc0,0x0000000,0x6800050,0x400044e,0xc7fea40,0x40023cf,
+	0x680400f,0x780718e,0x540338e,0x200e38f,0x4001c4e,0x7000c40,0xa2140c0,0xc0724c0,
+	0x7c031d0,0xcc00000,0x0000000,0x78072f1,0x9005031,0x90a2071,0xaa10140,0xaa0c001,
+	0xc074a80,0x0000000,0x649cf00,0xa219000,0xc072180,0x7c031d0,0xc077a40,0x0000000,
+	0xc780140,0xc077980,0x0000000,0xc071fc0,0x7c031d0,0xcc00000,0x0000000,0x0000000,
 	0x0000000
 };
diff --git a/drivers/amlogic/amports/m6tvd/ucode/h264/h264mmc_linux.h b/drivers/amlogic/amports/m6tvd/ucode/h264/h264mmc_linux.h
index d58868f10871..5ffde3db98d7 100755
--- a/drivers/amlogic/amports/m6tvd/ucode/h264/h264mmc_linux.h
+++ b/drivers/amlogic/amports/m6tvd/ucode/h264/h264mmc_linux.h
@@ -5,9 +5,9 @@ const u32 MicroCode[] __attribute__ ((aligned (8))) = {
 	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
 	0xcc00000,0x0400000,0xcc00000,0x0400000,0xcc00000,0x0400000,0xcc00000,0x0400000,
 	0xcc00000,0x0400000,0xc780200,0x6431e07,0xcc00000,0x0400000,0xcc00000,0x0400000,
-	0xcc00000,0x0400000,0x9027207,0xa6140c8,0x0000000,0xc00cec0,0x0000000,0xc780880,
-	0xa614148,0x0000000,0xc003800,0x0000000,0xc780740,0xa614188,0x0000000,0xc006840,
-	0x0000000,0xc780600,0xa6141c8,0x0000000,0xc007000,0x0000000,0xc7804c0,0xa614208,
+	0xcc00000,0x0400000,0x9027207,0xa6140c8,0x0000000,0xc00ce80,0x0000000,0xc780880,
+	0xa614148,0x0000000,0xc003800,0x0000000,0xc780740,0xa614188,0x0000000,0xc006800,
+	0x0000000,0xc780600,0xa6141c8,0x0000000,0xc006fc0,0x0000000,0xc7804c0,0xa614208,
 	0x0000000,0xc000c40,0x0000000,0xc780380,0xa62c288,0x6431f20,0x6433f21,0xe000860,
 	0x0000000,0x0000000,0xf000800,0x6031f20,0xf010840,0xc780100,0x6033f21,0xc780080,
 	0x0000000,0x8031e00,0xcc00000,0x0400000,0x78077c5,0x4002c4e,0x7400c45,0x7807800,
@@ -48,84 +48,84 @@ const u32 MicroCode[] __attribute__ ((aligned (8))) = {
 	0x7400070,0xcc00000,0x0000000,0x680400f,0x5403c4d,0x2031c4f,0x4001c71,0x7000c40,
 	0x6800081,0x9542001,0x7400c40,0x4401c31,0x7000c04,0x649cf30,0xa20c830,0x5808044,
 	0xbe20041,0x7801c01,0xaa10001,0xa216004,0x6800001,0x7c01c01,0x5809040,0xa218041,
-	0x6800001,0x9523001,0x7400c40,0xc781340,0x44012cb,0xc07f400,0x0000000,0x649c001,
-	0xbef7fc1,0x6800001,0x940504d,0x9184c00,0xaa243f0,0x68001c4,0xaa1c030,0x6800004,
-	0xb614170,0x4404130,0xb60c0f0,0x4002130,0x68001c4,0x94a3044,0x7802130,0xaa0c030,
-	0x0000000,0x9801201,0x90c2c00,0xa60c0b0,0x9101c00,0x9801201,0x9521070,0x4401c71,
-	0x7000c70,0x90a2c30,0x9542070,0x78017f0,0x9583070,0x6803230,0x2030370,0x7000c30,
-	0x9610070,0x6800030,0x2030370,0x7000c04,0x5402c0b,0x20302f0,0x6836eb1,0x2031c31,
-	0xcb80031,0x0000000,0x609ca04,0x0000000,0x0000000,0xc7806c0,0x609c101,0x649ca30,
-	0x9610c04,0x609ca30,0xc780580,0x609c201,0x609cb04,0x0000000,0x0000000,0xc780440,
-	0x609c301,0x649cb30,0x9610c04,0x609cb30,0xc780300,0x609c401,0x609cc04,0x0000000,
-	0x0000000,0xc7801c0,0x609c501,0x649cc30,0x9610c04,0x609cc30,0xc780080,0x609c601,
-	0x649cf01,0xbe09001,0xbe100c0,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x0000000,
-	0x680000b,0x7803d0d,0xa21880d,0x780318c,0x7801b8d,0xb00334c,0x0000000,0x7801b8c,
-	0x78071cd,0x649cf00,0xbe0d000,0xb01230d,0xc7800c0,0x680004c,0xb80f30d,0x0000000,
-	0xc07d440,0x0000000,0xa61bfcd,0x0000000,0xc079e00,0x0000000,0xc7801c0,0x0000000,
-	0xc07dec0,0x0000000,0x40012cb,0xaa0818b,0xc7ff940,0xaa1400b,0x6800080,0x950800b,
-	0x609c000,0x609ce00,0xcc00000,0x0000000,0x680000b,0x78071cd,0xaa4800d,0x0000000,
-	0xc07ce40,0x0000000,0xa61bfcd,0x0000000,0xc079800,0x0000000,0xc780280,0x0000000,
-	0x7807180,0xa807340,0x0000000,0xc07d800,0x0000000,0x40012cb,0xaa0818b,0xc7ffb80,
-	0xaa1c00b,0x6800080,0x950800b,0x609c000,0x609ce00,0xc7ffa00,0x680000b,0xcc00000,
-	0x0000000,0x680400f,0x780704e,0x6800008,0x6800007,0x7807186,0x78072c0,0xa225800,
-	0x7c07087,0x5403c46,0x2031c4f,0x7000c70,0x9005c30,0x6800041,0x3401c01,0x7c07081,
-	0x5403c48,0x2031c4f,0x4001c71,0x7000c70,0xa23c0f0,0xa406188,0x6800000,0x9444c00,
-	0x7400c70,0xc780280,0x6800047,0x6800000,0x9406c00,0x7400c70,0x6810000,0xbc04030,
-	0x78071c0,0x4401000,0x7c071c0,0x4001208,0xb8ec388,0xcc00000,0x7c07207,0x680400f,
-	0x780704e,0x6800008,0x6800007,0x7807186,0x5403c48,0x2031c4f,0x4001c71,0x7000c70,
-	0x6810000,0xa007030,0xa805188,0x0000000,0x9c302b0,0xc7800c0,0x7400c70,0x6800047,
-	0x4001208,0xb8f3388,0xcc00000,0x7c071c7,0x400138e,0x780734d,0x240e38d,0x7807300,
-	0xb610080,0x0000000,0x200e34e,0x400138e,0x6800008,0x7807109,0x680a00a,0xa403209,
-	0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,
-	0x9610001,0xa42a380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,
-	0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa18080,0x6800000,0xc780180,0x94421c0,
-	0xc780100,0x9c07087,0xc780080,0x9c070c7,0xbe4c307,0x6800000,0x94021c0,0x7807200,
-	0x4401000,0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,
-	0x3000040,0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0x649cf00,
-	0xbe14080,0x7400187,0xbe0d000,0x4001208,0xb8ce248,0xcc00000,0x0000000,0x7807108,
-	0x7807149,0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,
-	0x7000c40,0x4001c71,0x7000c41,0x9610001,0xa42a380,0x7000105,0x6804031,0x5403005,
-	0x2006031,0x4001186,0x7000187,0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa1c080,
-	0x6800000,0xc780200,0x94441c0,0x9c07087,0xc780140,0x9c07107,0x9c070c7,0xc780080,
-	0x9c07147,0xbe4c307,0x6800000,0x94021c0,0x7807200,0x4401000,0x7c07200,0x4401046,
-	0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,0x7c07080,0x6810000,
-	0xbc04007,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x7400187,0x4001208,0xb8ce248,
-	0xcc00000,0x0000000,0x680000d,0x7807300,0xb60c080,0x0000000,0x680004d,0x7807108,
-	0x7807149,0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,
-	0x7000c40,0x4001c71,0x7000c41,0x9610001,0x3800340,0xa41c380,0x7000105,0x6804031,
-	0x5403005,0x2006031,0x4001186,0x7000187,0xa24c307,0x6800000,0x94061c0,0x7807200,
-	0x4401000,0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,
-	0x3000040,0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0xcc00000,
-	0x7400187,0x4001208,0xb8db248,0xcc00000,0x0000000,0x400134d,0x780734c,0x240d34c,
-	0x7807300,0xb610080,0x0000000,0x200d30d,0x400134d,0x6800008,0x7807109,0x680a00a,
-	0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,
-	0x7000c41,0x9610001,0xa417340,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,
-	0x7000187,0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa18080,0x68000c0,0xc780180,
-	0x94821c0,0xc780100,0x9807107,0xc780080,0x9807147,0x4006c46,0x7400c4e,0xcc00000,
-	0x7400187,0x4001208,0xb8e1248,0xcc00000,0x0000000,0x440138e,0x680000d,0x7807300,
-	0xb60c080,0x0000000,0x680004d,0x7807108,0x7807149,0x680a00a,0xa403209,0xcc00000,
+	0x6800001,0x9523001,0x7400c40,0xc781300,0x44012cb,0xc07f400,0x0000000,0x649c001,
+	0xbef7fc1,0x9101044,0x5409041,0x940504d,0x9184c00,0xaa243f0,0x68001c4,0xaa1c030,
+	0x6800004,0xb614170,0x4404130,0xb60c0f0,0x4002130,0x68001c4,0x94a3044,0x7802130,
+	0xaa0c030,0x0000000,0x9801201,0xa60c0b0,0x90c2c00,0x9801201,0x4401c71,0x7000c70,
+	0x90a2c30,0x9542070,0x78017f0,0x9583070,0x6803230,0x2030370,0x7000c30,0x9610070,
+	0x6800030,0x2030370,0x7000c04,0x5402c0b,0x20302f0,0x6836e71,0x2031c31,0xcb80031,
+	0x0000000,0x609ca04,0x0000000,0x0000000,0xc7806c0,0x609c101,0x649ca30,0x9610c04,
+	0x609ca30,0xc780580,0x609c201,0x609cb04,0x0000000,0x0000000,0xc780440,0x609c301,
+	0x649cb30,0x9610c04,0x609cb30,0xc780300,0x609c401,0x609cc04,0x0000000,0x0000000,
+	0xc7801c0,0x609c501,0x649cc30,0x9610c04,0x609cc30,0xc780080,0x609c601,0x649cf01,
+	0xbe09001,0xbe100c0,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x0000000,0x680000b,
+	0x7803d0d,0xa21880d,0x780318c,0x7801b8d,0xb00334c,0x0000000,0x7801b8c,0x78071cd,
+	0x649cf00,0xbe0d000,0xb01230d,0xc7800c0,0x680004c,0xb80f30d,0x0000000,0xc07d480,
+	0x0000000,0xa61bfcd,0x0000000,0xc079e40,0x0000000,0xc7801c0,0x0000000,0xc07df00,
+	0x0000000,0x40012cb,0xaa0818b,0xc7ff940,0xaa1400b,0x6800080,0x950800b,0x609c000,
+	0x609ce00,0xcc00000,0x0000000,0x680000b,0x78071cd,0xaa4800d,0x0000000,0xc07ce80,
+	0x0000000,0xa61bfcd,0x0000000,0xc079840,0x0000000,0xc780280,0x0000000,0x7807180,
+	0xa807340,0x0000000,0xc07d840,0x0000000,0x40012cb,0xaa0818b,0xc7ffb80,0xaa1c00b,
+	0x6800080,0x950800b,0x609c000,0x609ce00,0xc7ffa00,0x680000b,0xcc00000,0x0000000,
+	0x680400f,0x780704e,0x6800008,0x6800007,0x7807186,0x78072c0,0xa225800,0x7c07087,
+	0x5403c46,0x2031c4f,0x7000c70,0x9005c30,0x6800041,0x3401c01,0x7c07081,0x5403c48,
+	0x2031c4f,0x4001c71,0x7000c70,0xa23c0f0,0xa406188,0x6800000,0x9444c00,0x7400c70,
+	0xc780280,0x6800047,0x6800000,0x9406c00,0x7400c70,0x6810000,0xbc04030,0x78071c0,
+	0x4401000,0x7c071c0,0x4001208,0xb8ec388,0xcc00000,0x7c07207,0x680400f,0x780704e,
+	0x6800008,0x6800007,0x7807186,0x5403c48,0x2031c4f,0x4001c71,0x7000c70,0x6810000,
+	0xa007030,0xa805188,0x0000000,0x9c302b0,0xc7800c0,0x7400c70,0x6800047,0x4001208,
+	0xb8f3388,0xcc00000,0x7c071c7,0x400138e,0x780734d,0x240e38d,0x7807300,0xb610080,
+	0x0000000,0x200e34e,0x400138e,0x6800008,0x7807109,0x680a00a,0xa403209,0xcc00000,
 	0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,0x9610001,
-	0x3800340,0xb01b380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,
-	0xa250307,0x6800000,0x94061c0,0x7807200,0x4401000,0x7c07200,0x4401046,0x7000040,
+	0xa42a380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,0x4001c44,
+	0x7000c70,0x9002030,0xaa14040,0xaa18080,0x6800000,0xc780180,0x94421c0,0xc780100,
+	0x9c07087,0xc780080,0x9c070c7,0xbe4c307,0x6800000,0x94021c0,0x7807200,0x4401000,
+	0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,
+	0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0x649cf00,0xbe14080,
+	0x7400187,0xbe0d000,0x4001208,0xb8ce248,0xcc00000,0x0000000,0x7807108,0x7807149,
+	0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,
+	0x4001c71,0x7000c41,0x9610001,0xa42a380,0x7000105,0x6804031,0x5403005,0x2006031,
+	0x4001186,0x7000187,0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa1c080,0x6800000,
+	0xc780200,0x94441c0,0x9c07087,0xc780140,0x9c07107,0x9c070c7,0xc780080,0x9c07147,
+	0xbe4c307,0x6800000,0x94021c0,0x7807200,0x4401000,0x7c07200,0x4401046,0x7000040,
 	0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,0x7c07080,0x6810000,0xbc04007,
-	0x78071c0,0x4401000,0x7c071c0,0x7400187,0x4001208,0xb8dc248,0xcc00000,0x0000000,
-	0x6800010,0x6800011,0x78072f1,0x9005031,0xa638140,0x0000000,0xc07aac0,0x0000000,
-	0x780e100,0xbe18080,0x9001400,0xc07a240,0x0000000,0xc781900,0x0000000,0xc07b200,
-	0x0000000,0xc781800,0x6800000,0x90a2071,0x780e130,0xa85c001,0xa85b030,0x680e10f,
-	0x70003c0,0xaa2c040,0xaa40080,0xaa540c0,0xaa78100,0x0000000,0x4605040,0xc500840,
-	0x0000000,0x4606040,0xc5011c0,0xc781380,0x4001c4f,0x7000c4e,0xc07b180,0x0000000,
-	0xc7ffc00,0x40023cf,0x4001c4f,0x7000c4e,0xc07c0c0,0x0000000,0xc7ffa80,0x40023cf,
-	0x4002c4f,0x7000c4e,0xc07ce00,0x0000000,0x4001c4f,0x7000c4d,0xc07d9c0,0x0000000,
-	0xc7ff800,0x40033cf,0x4001c4f,0x7000c4e,0xc07e440,0x0000000,0xc7ff680,0x40023cf,
-	0xc079c40,0x0000000,0xc079480,0x0000000,0x6800007,0x7807331,0xb250071,0x7807770,
-	0x7807780,0x9610c00,0x7807841,0x7807880,0x9610040,0x2401c01,0x7c07841,0x5810001,
-	0x7c07880,0x78077c1,0x7807800,0x9610040,0x2401c01,0x7c077c1,0x5810001,0x7c07800,
-	0xc7801c0,0x7c07441,0x7c07447,0x7c077c7,0x7c07807,0x7c07847,0x7c07887,0x7c07347,
-	0x7c07487,0x7c07647,0x7c07687,0x7c076c7,0x7c07707,0x7c07747,0x7c07787,0xc7fec40,
-	0x40013cf,0x4001c4f,0x7000c4e,0xc07bfc0,0x0000000,0x6800050,0x400044e,0xc7fea40,
-	0x40023cf,0x680400f,0x780718e,0x540338e,0x200e38f,0x4001c4e,0x7000c40,0xa2140c0,
-	0xc072480,0x7c031d0,0xcc00000,0x0000000,0x78072f1,0x9005031,0x90a2071,0xaa10140,
-	0xaa0c001,0xc074a40,0x0000000,0x649cf00,0xa219000,0xc072140,0x7c031d0,0xc077a40,
-	0x0000000,0xc780140,0xc077980,0x0000000,0xc071f80,0x7c031d0,0xcc00000,0x0000000,
+	0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x7400187,0x4001208,0xb8ce248,0xcc00000,
+	0x0000000,0x680000d,0x7807300,0xb60c080,0x0000000,0x680004d,0x7807108,0x7807149,
+	0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,
+	0x4001c71,0x7000c41,0x9610001,0x3800340,0xa41c380,0x7000105,0x6804031,0x5403005,
+	0x2006031,0x4001186,0x7000187,0xa24c307,0x6800000,0x94061c0,0x7807200,0x4401000,
+	0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,
+	0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x7400187,
+	0x4001208,0xb8db248,0xcc00000,0x0000000,0x400134d,0x780734c,0x240d34c,0x7807300,
+	0xb610080,0x0000000,0x200d30d,0x400134d,0x6800008,0x7807109,0x680a00a,0xa403209,
+	0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,
+	0x9610001,0xa417340,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,
+	0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa18080,0x68000c0,0xc780180,0x94821c0,
+	0xc780100,0x9807107,0xc780080,0x9807147,0x4006c46,0x7400c4e,0xcc00000,0x7400187,
+	0x4001208,0xb8e1248,0xcc00000,0x0000000,0x440138e,0x680000d,0x7807300,0xb60c080,
+	0x0000000,0x680004d,0x7807108,0x7807149,0x680a00a,0xa403209,0xcc00000,0x0000000,
+	0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,0x9610001,0x3800340,
+	0xb01b380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,0xa250307,
+	0x6800000,0x94061c0,0x7807200,0x4401000,0x7c07200,0x4401046,0x7000040,0x9005000,
+	0x6800041,0x3401001,0x7807080,0x3000040,0x7c07080,0x6810000,0xbc04007,0x78071c0,
+	0x4401000,0x7c071c0,0x7400187,0x4001208,0xb8dc248,0xcc00000,0x0000000,0x6800010,
+	0x6800011,0x78072f1,0x9005031,0xa638140,0x0000000,0xc07aac0,0x0000000,0x780e100,
+	0xbe18080,0x9001400,0xc07a240,0x0000000,0xc781900,0x0000000,0xc07b200,0x0000000,
+	0xc781800,0x6800000,0x90a2071,0x780e130,0xa85c001,0xa85b030,0x680e10f,0x70003c0,
+	0xaa2c040,0xaa40080,0xaa540c0,0xaa78100,0x0000000,0x4605040,0xc500840,0x0000000,
+	0x4606040,0xc5011c0,0xc781380,0x4001c4f,0x7000c4e,0xc07b180,0x0000000,0xc7ffc00,
+	0x40023cf,0x4001c4f,0x7000c4e,0xc07c0c0,0x0000000,0xc7ffa80,0x40023cf,0x4002c4f,
+	0x7000c4e,0xc07ce00,0x0000000,0x4001c4f,0x7000c4d,0xc07d9c0,0x0000000,0xc7ff800,
+	0x40033cf,0x4001c4f,0x7000c4e,0xc07e440,0x0000000,0xc7ff680,0x40023cf,0xc079c40,
+	0x0000000,0xc079480,0x0000000,0x6800007,0x7807331,0xb250071,0x7807770,0x7807780,
+	0x9610c00,0x7807841,0x7807880,0x9610040,0x2401c01,0x7c07841,0x5810001,0x7c07880,
+	0x78077c1,0x7807800,0x9610040,0x2401c01,0x7c077c1,0x5810001,0x7c07800,0xc7801c0,
+	0x7c07441,0x7c07447,0x7c077c7,0x7c07807,0x7c07847,0x7c07887,0x7c07347,0x7c07487,
+	0x7c07647,0x7c07687,0x7c076c7,0x7c07707,0x7c07747,0x7c07787,0xc7fec40,0x40013cf,
+	0x4001c4f,0x7000c4e,0xc07bfc0,0x0000000,0x6800050,0x400044e,0xc7fea40,0x40023cf,
+	0x680400f,0x780718e,0x540338e,0x200e38f,0x4001c4e,0x7000c40,0xa2140c0,0xc0724c0,
+	0x7c031d0,0xcc00000,0x0000000,0x78072f1,0x9005031,0x90a2071,0xaa10140,0xaa0c001,
+	0xc074a80,0x0000000,0x649cf00,0xa219000,0xc072180,0x7c031d0,0xc077a40,0x0000000,
+	0xc780140,0xc077980,0x0000000,0xc071fc0,0x7c031d0,0xcc00000,0x0000000,0x0000000,
 	0x0000000
 };
diff --git a/drivers/amlogic/amports/m8/ucode/h264/h264mmc_linux.h b/drivers/amlogic/amports/m8/ucode/h264/h264mmc_linux.h
index d58868f10871..5ffde3db98d7 100755
--- a/drivers/amlogic/amports/m8/ucode/h264/h264mmc_linux.h
+++ b/drivers/amlogic/amports/m8/ucode/h264/h264mmc_linux.h
@@ -5,9 +5,9 @@ const u32 MicroCode[] __attribute__ ((aligned (8))) = {
 	0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,0x0000000,
 	0xcc00000,0x0400000,0xcc00000,0x0400000,0xcc00000,0x0400000,0xcc00000,0x0400000,
 	0xcc00000,0x0400000,0xc780200,0x6431e07,0xcc00000,0x0400000,0xcc00000,0x0400000,
-	0xcc00000,0x0400000,0x9027207,0xa6140c8,0x0000000,0xc00cec0,0x0000000,0xc780880,
-	0xa614148,0x0000000,0xc003800,0x0000000,0xc780740,0xa614188,0x0000000,0xc006840,
-	0x0000000,0xc780600,0xa6141c8,0x0000000,0xc007000,0x0000000,0xc7804c0,0xa614208,
+	0xcc00000,0x0400000,0x9027207,0xa6140c8,0x0000000,0xc00ce80,0x0000000,0xc780880,
+	0xa614148,0x0000000,0xc003800,0x0000000,0xc780740,0xa614188,0x0000000,0xc006800,
+	0x0000000,0xc780600,0xa6141c8,0x0000000,0xc006fc0,0x0000000,0xc7804c0,0xa614208,
 	0x0000000,0xc000c40,0x0000000,0xc780380,0xa62c288,0x6431f20,0x6433f21,0xe000860,
 	0x0000000,0x0000000,0xf000800,0x6031f20,0xf010840,0xc780100,0x6033f21,0xc780080,
 	0x0000000,0x8031e00,0xcc00000,0x0400000,0x78077c5,0x4002c4e,0x7400c45,0x7807800,
@@ -48,84 +48,84 @@ const u32 MicroCode[] __attribute__ ((aligned (8))) = {
 	0x7400070,0xcc00000,0x0000000,0x680400f,0x5403c4d,0x2031c4f,0x4001c71,0x7000c40,
 	0x6800081,0x9542001,0x7400c40,0x4401c31,0x7000c04,0x649cf30,0xa20c830,0x5808044,
 	0xbe20041,0x7801c01,0xaa10001,0xa216004,0x6800001,0x7c01c01,0x5809040,0xa218041,
-	0x6800001,0x9523001,0x7400c40,0xc781340,0x44012cb,0xc07f400,0x0000000,0x649c001,
-	0xbef7fc1,0x6800001,0x940504d,0x9184c00,0xaa243f0,0x68001c4,0xaa1c030,0x6800004,
-	0xb614170,0x4404130,0xb60c0f0,0x4002130,0x68001c4,0x94a3044,0x7802130,0xaa0c030,
-	0x0000000,0x9801201,0x90c2c00,0xa60c0b0,0x9101c00,0x9801201,0x9521070,0x4401c71,
-	0x7000c70,0x90a2c30,0x9542070,0x78017f0,0x9583070,0x6803230,0x2030370,0x7000c30,
-	0x9610070,0x6800030,0x2030370,0x7000c04,0x5402c0b,0x20302f0,0x6836eb1,0x2031c31,
-	0xcb80031,0x0000000,0x609ca04,0x0000000,0x0000000,0xc7806c0,0x609c101,0x649ca30,
-	0x9610c04,0x609ca30,0xc780580,0x609c201,0x609cb04,0x0000000,0x0000000,0xc780440,
-	0x609c301,0x649cb30,0x9610c04,0x609cb30,0xc780300,0x609c401,0x609cc04,0x0000000,
-	0x0000000,0xc7801c0,0x609c501,0x649cc30,0x9610c04,0x609cc30,0xc780080,0x609c601,
-	0x649cf01,0xbe09001,0xbe100c0,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x0000000,
-	0x680000b,0x7803d0d,0xa21880d,0x780318c,0x7801b8d,0xb00334c,0x0000000,0x7801b8c,
-	0x78071cd,0x649cf00,0xbe0d000,0xb01230d,0xc7800c0,0x680004c,0xb80f30d,0x0000000,
-	0xc07d440,0x0000000,0xa61bfcd,0x0000000,0xc079e00,0x0000000,0xc7801c0,0x0000000,
-	0xc07dec0,0x0000000,0x40012cb,0xaa0818b,0xc7ff940,0xaa1400b,0x6800080,0x950800b,
-	0x609c000,0x609ce00,0xcc00000,0x0000000,0x680000b,0x78071cd,0xaa4800d,0x0000000,
-	0xc07ce40,0x0000000,0xa61bfcd,0x0000000,0xc079800,0x0000000,0xc780280,0x0000000,
-	0x7807180,0xa807340,0x0000000,0xc07d800,0x0000000,0x40012cb,0xaa0818b,0xc7ffb80,
-	0xaa1c00b,0x6800080,0x950800b,0x609c000,0x609ce00,0xc7ffa00,0x680000b,0xcc00000,
-	0x0000000,0x680400f,0x780704e,0x6800008,0x6800007,0x7807186,0x78072c0,0xa225800,
-	0x7c07087,0x5403c46,0x2031c4f,0x7000c70,0x9005c30,0x6800041,0x3401c01,0x7c07081,
-	0x5403c48,0x2031c4f,0x4001c71,0x7000c70,0xa23c0f0,0xa406188,0x6800000,0x9444c00,
-	0x7400c70,0xc780280,0x6800047,0x6800000,0x9406c00,0x7400c70,0x6810000,0xbc04030,
-	0x78071c0,0x4401000,0x7c071c0,0x4001208,0xb8ec388,0xcc00000,0x7c07207,0x680400f,
-	0x780704e,0x6800008,0x6800007,0x7807186,0x5403c48,0x2031c4f,0x4001c71,0x7000c70,
-	0x6810000,0xa007030,0xa805188,0x0000000,0x9c302b0,0xc7800c0,0x7400c70,0x6800047,
-	0x4001208,0xb8f3388,0xcc00000,0x7c071c7,0x400138e,0x780734d,0x240e38d,0x7807300,
-	0xb610080,0x0000000,0x200e34e,0x400138e,0x6800008,0x7807109,0x680a00a,0xa403209,
-	0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,
-	0x9610001,0xa42a380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,
-	0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa18080,0x6800000,0xc780180,0x94421c0,
-	0xc780100,0x9c07087,0xc780080,0x9c070c7,0xbe4c307,0x6800000,0x94021c0,0x7807200,
-	0x4401000,0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,
-	0x3000040,0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0x649cf00,
-	0xbe14080,0x7400187,0xbe0d000,0x4001208,0xb8ce248,0xcc00000,0x0000000,0x7807108,
-	0x7807149,0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,
-	0x7000c40,0x4001c71,0x7000c41,0x9610001,0xa42a380,0x7000105,0x6804031,0x5403005,
-	0x2006031,0x4001186,0x7000187,0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa1c080,
-	0x6800000,0xc780200,0x94441c0,0x9c07087,0xc780140,0x9c07107,0x9c070c7,0xc780080,
-	0x9c07147,0xbe4c307,0x6800000,0x94021c0,0x7807200,0x4401000,0x7c07200,0x4401046,
-	0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,0x7c07080,0x6810000,
-	0xbc04007,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x7400187,0x4001208,0xb8ce248,
-	0xcc00000,0x0000000,0x680000d,0x7807300,0xb60c080,0x0000000,0x680004d,0x7807108,
-	0x7807149,0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,
-	0x7000c40,0x4001c71,0x7000c41,0x9610001,0x3800340,0xa41c380,0x7000105,0x6804031,
-	0x5403005,0x2006031,0x4001186,0x7000187,0xa24c307,0x6800000,0x94061c0,0x7807200,
-	0x4401000,0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,
-	0x3000040,0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0xcc00000,
-	0x7400187,0x4001208,0xb8db248,0xcc00000,0x0000000,0x400134d,0x780734c,0x240d34c,
-	0x7807300,0xb610080,0x0000000,0x200d30d,0x400134d,0x6800008,0x7807109,0x680a00a,
-	0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,
-	0x7000c41,0x9610001,0xa417340,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,
-	0x7000187,0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa18080,0x68000c0,0xc780180,
-	0x94821c0,0xc780100,0x9807107,0xc780080,0x9807147,0x4006c46,0x7400c4e,0xcc00000,
-	0x7400187,0x4001208,0xb8e1248,0xcc00000,0x0000000,0x440138e,0x680000d,0x7807300,
-	0xb60c080,0x0000000,0x680004d,0x7807108,0x7807149,0x680a00a,0xa403209,0xcc00000,
+	0x6800001,0x9523001,0x7400c40,0xc781300,0x44012cb,0xc07f400,0x0000000,0x649c001,
+	0xbef7fc1,0x9101044,0x5409041,0x940504d,0x9184c00,0xaa243f0,0x68001c4,0xaa1c030,
+	0x6800004,0xb614170,0x4404130,0xb60c0f0,0x4002130,0x68001c4,0x94a3044,0x7802130,
+	0xaa0c030,0x0000000,0x9801201,0xa60c0b0,0x90c2c00,0x9801201,0x4401c71,0x7000c70,
+	0x90a2c30,0x9542070,0x78017f0,0x9583070,0x6803230,0x2030370,0x7000c30,0x9610070,
+	0x6800030,0x2030370,0x7000c04,0x5402c0b,0x20302f0,0x6836e71,0x2031c31,0xcb80031,
+	0x0000000,0x609ca04,0x0000000,0x0000000,0xc7806c0,0x609c101,0x649ca30,0x9610c04,
+	0x609ca30,0xc780580,0x609c201,0x609cb04,0x0000000,0x0000000,0xc780440,0x609c301,
+	0x649cb30,0x9610c04,0x609cb30,0xc780300,0x609c401,0x609cc04,0x0000000,0x0000000,
+	0xc7801c0,0x609c501,0x649cc30,0x9610c04,0x609cc30,0xc780080,0x609c601,0x649cf01,
+	0xbe09001,0xbe100c0,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x0000000,0x680000b,
+	0x7803d0d,0xa21880d,0x780318c,0x7801b8d,0xb00334c,0x0000000,0x7801b8c,0x78071cd,
+	0x649cf00,0xbe0d000,0xb01230d,0xc7800c0,0x680004c,0xb80f30d,0x0000000,0xc07d480,
+	0x0000000,0xa61bfcd,0x0000000,0xc079e40,0x0000000,0xc7801c0,0x0000000,0xc07df00,
+	0x0000000,0x40012cb,0xaa0818b,0xc7ff940,0xaa1400b,0x6800080,0x950800b,0x609c000,
+	0x609ce00,0xcc00000,0x0000000,0x680000b,0x78071cd,0xaa4800d,0x0000000,0xc07ce80,
+	0x0000000,0xa61bfcd,0x0000000,0xc079840,0x0000000,0xc780280,0x0000000,0x7807180,
+	0xa807340,0x0000000,0xc07d840,0x0000000,0x40012cb,0xaa0818b,0xc7ffb80,0xaa1c00b,
+	0x6800080,0x950800b,0x609c000,0x609ce00,0xc7ffa00,0x680000b,0xcc00000,0x0000000,
+	0x680400f,0x780704e,0x6800008,0x6800007,0x7807186,0x78072c0,0xa225800,0x7c07087,
+	0x5403c46,0x2031c4f,0x7000c70,0x9005c30,0x6800041,0x3401c01,0x7c07081,0x5403c48,
+	0x2031c4f,0x4001c71,0x7000c70,0xa23c0f0,0xa406188,0x6800000,0x9444c00,0x7400c70,
+	0xc780280,0x6800047,0x6800000,0x9406c00,0x7400c70,0x6810000,0xbc04030,0x78071c0,
+	0x4401000,0x7c071c0,0x4001208,0xb8ec388,0xcc00000,0x7c07207,0x680400f,0x780704e,
+	0x6800008,0x6800007,0x7807186,0x5403c48,0x2031c4f,0x4001c71,0x7000c70,0x6810000,
+	0xa007030,0xa805188,0x0000000,0x9c302b0,0xc7800c0,0x7400c70,0x6800047,0x4001208,
+	0xb8f3388,0xcc00000,0x7c071c7,0x400138e,0x780734d,0x240e38d,0x7807300,0xb610080,
+	0x0000000,0x200e34e,0x400138e,0x6800008,0x7807109,0x680a00a,0xa403209,0xcc00000,
 	0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,0x9610001,
-	0x3800340,0xb01b380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,
-	0xa250307,0x6800000,0x94061c0,0x7807200,0x4401000,0x7c07200,0x4401046,0x7000040,
+	0xa42a380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,0x4001c44,
+	0x7000c70,0x9002030,0xaa14040,0xaa18080,0x6800000,0xc780180,0x94421c0,0xc780100,
+	0x9c07087,0xc780080,0x9c070c7,0xbe4c307,0x6800000,0x94021c0,0x7807200,0x4401000,
+	0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,
+	0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0x649cf00,0xbe14080,
+	0x7400187,0xbe0d000,0x4001208,0xb8ce248,0xcc00000,0x0000000,0x7807108,0x7807149,
+	0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,
+	0x4001c71,0x7000c41,0x9610001,0xa42a380,0x7000105,0x6804031,0x5403005,0x2006031,
+	0x4001186,0x7000187,0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa1c080,0x6800000,
+	0xc780200,0x94441c0,0x9c07087,0xc780140,0x9c07107,0x9c070c7,0xc780080,0x9c07147,
+	0xbe4c307,0x6800000,0x94021c0,0x7807200,0x4401000,0x7c07200,0x4401046,0x7000040,
 	0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,0x7c07080,0x6810000,0xbc04007,
-	0x78071c0,0x4401000,0x7c071c0,0x7400187,0x4001208,0xb8dc248,0xcc00000,0x0000000,
-	0x6800010,0x6800011,0x78072f1,0x9005031,0xa638140,0x0000000,0xc07aac0,0x0000000,
-	0x780e100,0xbe18080,0x9001400,0xc07a240,0x0000000,0xc781900,0x0000000,0xc07b200,
-	0x0000000,0xc781800,0x6800000,0x90a2071,0x780e130,0xa85c001,0xa85b030,0x680e10f,
-	0x70003c0,0xaa2c040,0xaa40080,0xaa540c0,0xaa78100,0x0000000,0x4605040,0xc500840,
-	0x0000000,0x4606040,0xc5011c0,0xc781380,0x4001c4f,0x7000c4e,0xc07b180,0x0000000,
-	0xc7ffc00,0x40023cf,0x4001c4f,0x7000c4e,0xc07c0c0,0x0000000,0xc7ffa80,0x40023cf,
-	0x4002c4f,0x7000c4e,0xc07ce00,0x0000000,0x4001c4f,0x7000c4d,0xc07d9c0,0x0000000,
-	0xc7ff800,0x40033cf,0x4001c4f,0x7000c4e,0xc07e440,0x0000000,0xc7ff680,0x40023cf,
-	0xc079c40,0x0000000,0xc079480,0x0000000,0x6800007,0x7807331,0xb250071,0x7807770,
-	0x7807780,0x9610c00,0x7807841,0x7807880,0x9610040,0x2401c01,0x7c07841,0x5810001,
-	0x7c07880,0x78077c1,0x7807800,0x9610040,0x2401c01,0x7c077c1,0x5810001,0x7c07800,
-	0xc7801c0,0x7c07441,0x7c07447,0x7c077c7,0x7c07807,0x7c07847,0x7c07887,0x7c07347,
-	0x7c07487,0x7c07647,0x7c07687,0x7c076c7,0x7c07707,0x7c07747,0x7c07787,0xc7fec40,
-	0x40013cf,0x4001c4f,0x7000c4e,0xc07bfc0,0x0000000,0x6800050,0x400044e,0xc7fea40,
-	0x40023cf,0x680400f,0x780718e,0x540338e,0x200e38f,0x4001c4e,0x7000c40,0xa2140c0,
-	0xc072480,0x7c031d0,0xcc00000,0x0000000,0x78072f1,0x9005031,0x90a2071,0xaa10140,
-	0xaa0c001,0xc074a40,0x0000000,0x649cf00,0xa219000,0xc072140,0x7c031d0,0xc077a40,
-	0x0000000,0xc780140,0xc077980,0x0000000,0xc071f80,0x7c031d0,0xcc00000,0x0000000,
+	0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x7400187,0x4001208,0xb8ce248,0xcc00000,
+	0x0000000,0x680000d,0x7807300,0xb60c080,0x0000000,0x680004d,0x7807108,0x7807149,
+	0x680a00a,0xa403209,0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,
+	0x4001c71,0x7000c41,0x9610001,0x3800340,0xa41c380,0x7000105,0x6804031,0x5403005,
+	0x2006031,0x4001186,0x7000187,0xa24c307,0x6800000,0x94061c0,0x7807200,0x4401000,
+	0x7c07200,0x4401046,0x7000040,0x9005000,0x6800041,0x3401001,0x7807080,0x3000040,
+	0x7c07080,0x6810000,0xbc04007,0x78071c0,0x4401000,0x7c071c0,0xcc00000,0x7400187,
+	0x4001208,0xb8db248,0xcc00000,0x0000000,0x400134d,0x780734c,0x240d34c,0x7807300,
+	0xb610080,0x0000000,0x200d30d,0x400134d,0x6800008,0x7807109,0x680a00a,0xa403209,
+	0xcc00000,0x0000000,0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,
+	0x9610001,0xa417340,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,
+	0x4001c44,0x7000c70,0x9002030,0xaa14040,0xaa18080,0x68000c0,0xc780180,0x94821c0,
+	0xc780100,0x9807107,0xc780080,0x9807147,0x4006c46,0x7400c4e,0xcc00000,0x7400187,
+	0x4001208,0xb8e1248,0xcc00000,0x0000000,0x440138e,0x680000d,0x7807300,0xb60c080,
+	0x0000000,0x680004d,0x7807108,0x7807149,0x680a00a,0xa403209,0xcc00000,0x0000000,
+	0x5402108,0x200410a,0x4002c44,0x7000c40,0x4001c71,0x7000c41,0x9610001,0x3800340,
+	0xb01b380,0x7000105,0x6804031,0x5403005,0x2006031,0x4001186,0x7000187,0xa250307,
+	0x6800000,0x94061c0,0x7807200,0x4401000,0x7c07200,0x4401046,0x7000040,0x9005000,
+	0x6800041,0x3401001,0x7807080,0x3000040,0x7c07080,0x6810000,0xbc04007,0x78071c0,
+	0x4401000,0x7c071c0,0x7400187,0x4001208,0xb8dc248,0xcc00000,0x0000000,0x6800010,
+	0x6800011,0x78072f1,0x9005031,0xa638140,0x0000000,0xc07aac0,0x0000000,0x780e100,
+	0xbe18080,0x9001400,0xc07a240,0x0000000,0xc781900,0x0000000,0xc07b200,0x0000000,
+	0xc781800,0x6800000,0x90a2071,0x780e130,0xa85c001,0xa85b030,0x680e10f,0x70003c0,
+	0xaa2c040,0xaa40080,0xaa540c0,0xaa78100,0x0000000,0x4605040,0xc500840,0x0000000,
+	0x4606040,0xc5011c0,0xc781380,0x4001c4f,0x7000c4e,0xc07b180,0x0000000,0xc7ffc00,
+	0x40023cf,0x4001c4f,0x7000c4e,0xc07c0c0,0x0000000,0xc7ffa80,0x40023cf,0x4002c4f,
+	0x7000c4e,0xc07ce00,0x0000000,0x4001c4f,0x7000c4d,0xc07d9c0,0x0000000,0xc7ff800,
+	0x40033cf,0x4001c4f,0x7000c4e,0xc07e440,0x0000000,0xc7ff680,0x40023cf,0xc079c40,
+	0x0000000,0xc079480,0x0000000,0x6800007,0x7807331,0xb250071,0x7807770,0x7807780,
+	0x9610c00,0x7807841,0x7807880,0x9610040,0x2401c01,0x7c07841,0x5810001,0x7c07880,
+	0x78077c1,0x7807800,0x9610040,0x2401c01,0x7c077c1,0x5810001,0x7c07800,0xc7801c0,
+	0x7c07441,0x7c07447,0x7c077c7,0x7c07807,0x7c07847,0x7c07887,0x7c07347,0x7c07487,
+	0x7c07647,0x7c07687,0x7c076c7,0x7c07707,0x7c07747,0x7c07787,0xc7fec40,0x40013cf,
+	0x4001c4f,0x7000c4e,0xc07bfc0,0x0000000,0x6800050,0x400044e,0xc7fea40,0x40023cf,
+	0x680400f,0x780718e,0x540338e,0x200e38f,0x4001c4e,0x7000c40,0xa2140c0,0xc0724c0,
+	0x7c031d0,0xcc00000,0x0000000,0x78072f1,0x9005031,0x90a2071,0xaa10140,0xaa0c001,
+	0xc074a80,0x0000000,0x649cf00,0xa219000,0xc072180,0x7c031d0,0xc077a40,0x0000000,
+	0xc780140,0xc077980,0x0000000,0xc071fc0,0x7c031d0,0xcc00000,0x0000000,0x0000000,
 	0x0000000
 };
diff --git a/drivers/amlogic/amports/vavs.c b/drivers/amlogic/amports/vavs.c
index 18718b111d2d..0e8b49bbb806 100755
--- a/drivers/amlogic/amports/vavs.c
+++ b/drivers/amlogic/amports/vavs.c
@@ -231,6 +231,7 @@ static void set_frame_info(vframe_t *vf, unsigned* duration)
         vf->ratio_control = (ar<<DISP_RATIO_ASPECT_RATIO_BIT);
         //vf->ratio_control |= DISP_RATIO_FORCECONFIG | DISP_RATIO_KEEPRATIO;
 
+    vf->flag = 0;
 }
 
 #ifdef HANDLE_AVS_IRQ
diff --git a/drivers/amlogic/amports/vh264.c b/drivers/amlogic/amports/vh264.c
index 3893ab959b98..a4aaf99a11d8 100755
--- a/drivers/amlogic/amports/vh264.c
+++ b/drivers/amlogic/amports/vh264.c
@@ -194,7 +194,6 @@ static u32 h264_ar;
 #ifdef DROP_B_FRAME_FOR_1080P_50_60FPS
 static u32 last_interlaced;
 #endif
-static u8 neg_poc_counter;
 static unsigned char h264_first_pts_ready;
 static unsigned char h264_first_valid_pts_ready;
 static u32 h264pts1, h264pts2;
@@ -211,9 +210,12 @@ static u32 vh264_running;
 static s32 vh264_stream_switching_state;
 static s32 vh264_eos;
 static struct vframe_s *p_last_vf;
+static u32 last_pts;
+static bool check_pts_discontinue;
 static u32 wait_buffer_counter;
 static uint error_recovery_mode = 0;
 static uint error_recovery_mode_in = 3;
+static uint error_recovery_mode_use = 3;
 
 static uint mb_total = 0, mb_width = 0,  mb_height=0;
 #define UCODE_IP_ONLY 2
@@ -475,6 +477,7 @@ static void set_frame_info(vframe_t *vf)
     vf->duration = frame_dur;
     vf->ratio_control = (min(h264_ar, (u32)DISP_RATIO_ASPECT_RATIO_MAX)) << DISP_RATIO_ASPECT_RATIO_BIT;
     vf->orientation = vh264_rotation;
+    vf->flag = 0;
 
 #ifdef CONFIG_POST_PROCESS_MANAGER_3D_PROCESS
     vf->trans_fmt = convert_3d_format(frame_packing_type);
@@ -923,7 +926,7 @@ static void vh264_isr(void)
         vh264_set_params();
 
     } else if ((cpu_cmd & 0xff) == 2) {
-        int frame_mb_only, pic_struct_present, pic_struct, prog_frame, poc_sel, idr_flag, neg_poc, eos;
+        int frame_mb_only, pic_struct_present, pic_struct, prog_frame, poc_sel, idr_flag, eos, error;
         int i, status, num_frame, b_offset;
         int current_error_count;
 
@@ -942,6 +945,11 @@ static void vh264_isr(void)
         for (i = 0 ; (i < num_frame) && (!vh264_eos) ; i++) {
             status = READ_VREG(AV_SCRATCH_1 + i);
             buffer_index = status & 0x1f;
+            error = status & 0x200;
+
+            if ((error_recovery_mode_use & 2) && error) {
+                check_pts_discontinue = true;
+            }
 
             if ((p_last_vf != NULL) && (p_last_vf->index == buffer_index)) {
                 continue;
@@ -955,7 +963,6 @@ static void vh264_isr(void)
             prog_frame = status & 0x100;
             poc_sel = status & 0x200;
             idr_flag = status & 0x400;
-            neg_poc = status & 0x800;
             frame_packing_type = (status >> 12) & 0x7;
             eos = (status >> 15) & 1;
 
@@ -1009,6 +1016,7 @@ static void vh264_isr(void)
                 pts_missed++;
 #endif
             }
+
             if (sync_outside == 0) {
                 if (h264_first_pts_ready == 0) {
                     if (pts_valid == 0) {
@@ -1071,17 +1079,9 @@ static void vh264_isr(void)
 
                 h264_pts_count++;
             } else {
-                if (idr_flag && pts_valid) {
-                    pts += neg_poc_counter * (frame_dur - (frame_dur >> 4));
-                } else {
+                if (!(idr_flag && pts_valid)) {
                     pts = 0;
                 }
-
-                if (neg_poc) {
-                    neg_poc_counter++;
-                } else {
-                    neg_poc_counter = 0;
-                }
             }
 
             if (timing_info_present_flag && frame_dur && use_idr_framerate) {
@@ -1105,6 +1105,22 @@ static void vh264_isr(void)
                 force_interlaced_frame = true;
             }
 
+            /* for frames with PTS, check if there is PTS discontinue based on previous frames (including error frames),
+             * force no VPTS discontinue reporting if we saw errors earlier but only once.
+             */
+            if ((pts_valid) && (check_pts_discontinue) && (!error)) {
+                if ((pts - last_pts) < 90000) {
+                    vf->flag = VFRAME_FLAG_NO_DISCONTINUE;
+                    check_pts_discontinue = false;
+                }
+            }
+
+            if (pts_valid) {
+                last_pts = pts;
+            } else {
+                last_pts += DUR2PTS(vf->duration);
+            }
+
             if (pic_struct_present) {
                 if ((pic_struct == PIC_TOP_BOT) || (pic_struct == PIC_BOT_TOP)) {
                     prog_frame = 0;
@@ -1135,11 +1151,15 @@ static void vh264_isr(void)
                 vf->canvas0Addr = vf->canvas1Addr = spec2canvas(&buffer_spec[buffer_index]);
                 vfbuf_use[buffer_index]++;
 
-                p_last_vf = vf;
+                if ((error_recovery_mode_use & 2) && error) {
+                    kfifo_put(&recycle_q, (const vframe_t **)&vf);
+                } else {
+                    p_last_vf = vf;
 
-                kfifo_put(&display_q, (const vframe_t **)&vf);
+                    kfifo_put(&display_q, (const vframe_t **)&vf);
 
-                vf_notify_receiver(PROVIDER_NAME,VFRAME_EVENT_PROVIDER_VFRAME_READY,NULL);
+                    vf_notify_receiver(PROVIDER_NAME,VFRAME_EVENT_PROVIDER_VFRAME_READY,NULL);
+                }
 
             } else {
                 if (pic_struct_present && pic_struct == PIC_TOP_BOT) {
@@ -1163,7 +1183,13 @@ static void vh264_isr(void)
                 vfbuf_use[buffer_index]++;
                 vf->ready_jiffies64=jiffies_64;
 
-                kfifo_put(&display_q, (const vframe_t **)&vf);
+                if ((error_recovery_mode_use & 2) && error) {
+                    kfifo_put(&recycle_q, (const vframe_t **)&vf);
+                    WRITE_VREG(AV_SCRATCH_0, 0);
+                    return IRQ_HANDLED;
+                } else {
+                    kfifo_put(&display_q, (const vframe_t **)&vf);
+                }
 
                 if (READ_VREG(AV_SCRATCH_F) & 2) {
                     vf_notify_receiver(PROVIDER_NAME,VFRAME_EVENT_PROVIDER_VFRAME_READY,NULL);
@@ -1439,10 +1465,9 @@ static void vh264_prot_init(void)
     WRITE_VREG(AV_SCRATCH_7, 0);
     WRITE_VREG(AV_SCRATCH_8, 0);
     WRITE_VREG(AV_SCRATCH_9, 0);
-	if(error_recovery_mode)/*if sysfs have set recovery mode,over write player intenel para settings.*/
-    	WRITE_VREG(AV_SCRATCH_F, (READ_VREG(AV_SCRATCH_F) & 0xffffffc3) | ((error_recovery_mode & 0x3) << 4));
-	else
-    	WRITE_VREG(AV_SCRATCH_F, (READ_VREG(AV_SCRATCH_F) & 0xffffffc3) | ((error_recovery_mode_in & 0x3) << 4));
+
+    error_recovery_mode_use = (error_recovery_mode != 0) ? error_recovery_mode : error_recovery_mode_in;
+    WRITE_VREG(AV_SCRATCH_F, (READ_VREG(AV_SCRATCH_F) & 0xffffffc3) | ((error_recovery_mode_use & 0x1) << 4));
 
     /* clear mailbox interrupt */
     WRITE_VREG(ASSIST_MBOX1_CLR_REG, 1);
@@ -1523,7 +1548,6 @@ static void vh264_local_init(void)
 #ifdef DROP_B_FRAME_FOR_1080P_50_60FPS
     last_interlaced = 1;
 #endif
-    neg_poc_counter = 0;
     h264_first_pts_ready = 0;
     h264_first_valid_pts_ready=0;
     h264pts1 = 0;
@@ -1533,6 +1557,8 @@ static void vh264_local_init(void)
     vh264_error_count = READ_VREG(AV_SCRATCH_D);
 
     p_last_vf = NULL;
+    check_pts_discontinue = false;
+    last_pts = 0;
     wait_buffer_counter = 0;
     vh264_no_disp_count = 0;
     fatal_error_flag = 0;
diff --git a/drivers/amlogic/amports/vh264_4k2k.c b/drivers/amlogic/amports/vh264_4k2k.c
index ee4b222b6238..17532b191a40 100755
--- a/drivers/amlogic/amports/vh264_4k2k.c
+++ b/drivers/amlogic/amports/vh264_4k2k.c
@@ -268,6 +268,7 @@ static void set_frame_info(vframe_t *vf)
 #endif
     vf->duration = frame_dur;
     vf->duration_pulldown = 0;
+    vf->flag = 0;
 
     ar = min(frame_ar, (u32)DISP_RATIO_ASPECT_RATIO_MAX);
     vf->ratio_control = (ar << DISP_RATIO_ASPECT_RATIO_BIT);
diff --git a/drivers/amlogic/amports/vh264mvc.c b/drivers/amlogic/amports/vh264mvc.c
index b8e3820ab1c4..32378e497aae 100755
--- a/drivers/amlogic/amports/vh264mvc.c
+++ b/drivers/amlogic/amports/vh264mvc.c
@@ -380,6 +380,7 @@ static vframe_t *vh264mvc_vf_get(void* op_arg)
 
         vf->width = frame_width;
         vf->height = frame_height;
+        vf->flag = 0;
     }
     return vf;
 
diff --git a/drivers/amlogic/amports/vh265.c b/drivers/amlogic/amports/vh265.c
index 27c2f1393ef6..5b163a78062e 100755
--- a/drivers/amlogic/amports/vh265.c
+++ b/drivers/amlogic/amports/vh265.c
@@ -2897,6 +2897,7 @@ static void set_frame_info(vframe_t *vf)
     vf->height = frame_height;
     vf->duration = frame_dur;
     vf->duration_pulldown = 0;
+    vf->flag = 0;
 
     ar = min(frame_ar, (u32)DISP_RATIO_ASPECT_RATIO_MAX);
     vf->ratio_control = (ar << DISP_RATIO_ASPECT_RATIO_BIT);
diff --git a/drivers/amlogic/amports/video.c b/drivers/amlogic/amports/video.c
index 30a215a1bc55..87b37b581203 100755
--- a/drivers/amlogic/amports/video.c
+++ b/drivers/amlogic/amports/video.c
@@ -1976,7 +1976,8 @@ static inline bool vpts_expire(vframe_t *cur_vf, vframe_t *next_vf)
     }
     /* check video PTS discontinuity */
     else if ((enable_video_discontinue_report) &&
-             (abs(systime - pts) > tsync_vpts_discontinuity_margin())) {
+             (abs(systime - pts) > tsync_vpts_discontinuity_margin()) &&
+             ((next_vf->flag & VFRAME_FLAG_NO_DISCONTINUE) == 0)) {
         pts = timestamp_vpts_get() + (cur_vf ? DUR2PTS(cur_vf->duration) : 0);
 			//printk("system=0x%x vpts=0x%x\n", systime, timestamp_vpts_get());
         if ((int)(systime - pts) >= 0){
diff --git a/drivers/amlogic/amports/vmjpeg.c b/drivers/amlogic/amports/vmjpeg.c
index 51eb68661835..de5fb9f85721 100755
--- a/drivers/amlogic/amports/vmjpeg.c
+++ b/drivers/amlogic/amports/vmjpeg.c
@@ -154,6 +154,7 @@ static void set_frame_info(vframe_t *vf)
     vf->duration = frame_dur;
     vf->ratio_control = 0;
     vf->duration_pulldown = 0;
+    vf->flag = 0;
 }
 
 static irqreturn_t vmjpeg_isr(int irq, void *dev_id)
diff --git a/drivers/amlogic/amports/vmpeg12.c b/drivers/amlogic/amports/vmpeg12.c
index 9c46d5cb803d..1d7cd38c8f24 100755
--- a/drivers/amlogic/amports/vmpeg12.c
+++ b/drivers/amlogic/amports/vmpeg12.c
@@ -189,6 +189,7 @@ static void set_frame_info(vframe_t *vf)
 
     vf->width  = frame_width = READ_VREG(MREG_PIC_WIDTH);
     vf->height = frame_height = READ_VREG(MREG_PIC_HEIGHT);
+    vf->flag = 0;
 
     if (frame_dur > 0) {
         vf->duration = frame_dur;
diff --git a/drivers/amlogic/amports/vmpeg4.c b/drivers/amlogic/amports/vmpeg4.c
index 7e1729673fdf..5e05d4a7a45d 100755
--- a/drivers/amlogic/amports/vmpeg4.c
+++ b/drivers/amlogic/amports/vmpeg4.c
@@ -403,6 +403,7 @@ static irqreturn_t vmpeg4_isr(int irq, void *dev_id)
             vf->width = vmpeg4_amstream_dec_info.width;
             vf->height = vmpeg4_amstream_dec_info.height;
             vf->bufWidth = 1920;
+            vf->flag = 0;
             vf->orientation = vmpeg4_rotation;
             vf->pts = pts;
             vf->pts_us64 = pts_us64;
@@ -431,6 +432,7 @@ static irqreturn_t vmpeg4_isr(int irq, void *dev_id)
             vf->width = vmpeg4_amstream_dec_info.width;
             vf->height = vmpeg4_amstream_dec_info.height;
             vf->bufWidth = 1920;
+            vf->flag = 0;
             vf->orientation = vmpeg4_rotation;
 
             vf->pts = 0;
@@ -466,6 +468,7 @@ static irqreturn_t vmpeg4_isr(int irq, void *dev_id)
             vf->width = vmpeg4_amstream_dec_info.width;
             vf->height = vmpeg4_amstream_dec_info.height;
             vf->bufWidth = 1920;
+            vf->flag = 0;
             vf->orientation = vmpeg4_rotation;
             vf->pts = pts;
             vf->pts_us64 = pts_us64;
diff --git a/drivers/amlogic/amports/vreal.c b/drivers/amlogic/amports/vreal.c
index c9ab83d446e4..5396d1701bd2 100755
--- a/drivers/amlogic/amports/vreal.c
+++ b/drivers/amlogic/amports/vreal.c
@@ -319,6 +319,7 @@ static irqreturn_t vreal_isr(int irq, void *dev_id)
         vf->width = info >> 16;
         vf->height = (info >> 4) & 0xfff;
         vf->bufWidth = 1920;
+        vf->flag = 0;
         vf->ratio_control = 0;
         set_aspect_ratio(vf, info & 0x0f);
         vf->duration_pulldown = 0;
diff --git a/drivers/amlogic/amports/vvc1.c b/drivers/amlogic/amports/vvc1.c
index 4b131b48bce9..224509260d9b 100755
--- a/drivers/amlogic/amports/vvc1.c
+++ b/drivers/amlogic/amports/vvc1.c
@@ -353,6 +353,7 @@ static irqreturn_t vvc1_isr(int irq, void *dev_id)
             vf->width = vvc1_amstream_dec_info.width;
             vf->height = vvc1_amstream_dec_info.height;
             vf->bufWidth = 1920;
+            vf->flag = 0;
 
             if (pts_valid) {
                 vf->pts = pts;
@@ -408,6 +409,7 @@ static irqreturn_t vvc1_isr(int irq, void *dev_id)
             vf->width = vvc1_amstream_dec_info.width;
             vf->height = vvc1_amstream_dec_info.height;
             vf->bufWidth = 1920;
+            vf->flag = 0;
 
             vf->pts = next_pts;
             vf->pts_us64 = next_pts_us64;
@@ -449,6 +451,7 @@ static irqreturn_t vvc1_isr(int irq, void *dev_id)
             vf->width = vvc1_amstream_dec_info.width;
             vf->height = vvc1_amstream_dec_info.height;
             vf->bufWidth = 1920;
+            vf->flag = 0;
 
             if (pts_valid) {
                 vf->pts = pts;
diff --git a/drivers/amlogic/deinterlace/deinterlace.c b/drivers/amlogic/deinterlace/deinterlace.c
index 2772cc5f81e7..5fed4b94b57f 100755
--- a/drivers/amlogic/deinterlace/deinterlace.c
+++ b/drivers/amlogic/deinterlace/deinterlace.c
@@ -4569,6 +4569,7 @@ static void process_vscale_skip(di_buf_t* di_buf, vframe_t* disp_vf)
             disp_vf->height = di_buf_i->vframe->height;
             disp_vf->duration = di_buf_i->vframe->duration;
             disp_vf->pts = di_buf_i->vframe->pts;
+            disp_vf->flag = di_buf_i->vframe->flag;
             disp_vf->canvas0Addr = di_post_buf0_canvas_idx[di_post_stru.canvas_id];
             disp_vf->canvas1Addr = di_post_buf0_canvas_idx[di_post_stru.canvas_id];
             canvas_config(di_post_buf0_canvas_idx[di_post_stru.canvas_id], di_buf_i->nr_adr, width*2, canvas_height, 0, 0);
diff --git a/drivers/amlogic/ppmgr/ppmgr_vpp.c b/drivers/amlogic/ppmgr/ppmgr_vpp.c
index 7e7630550d1b..25147a612bc1 100755
--- a/drivers/amlogic/ppmgr/ppmgr_vpp.c
+++ b/drivers/amlogic/ppmgr/ppmgr_vpp.c
@@ -1325,7 +1325,7 @@ static void process_vf_rotate(vframe_t *vf, ge2d_context_t *context, config_para
     new_vf->type = VIDTYPE_VIU_444 | VIDTYPE_VIU_SINGLE_PLANE | VIDTYPE_VIU_FIELD;
     new_vf->canvas0Addr = new_vf->canvas1Addr = index2canvas(pp_vf->index);
     new_vf->orientation = vf->orientation;
-
+    new_vf->flag = vf->flag;
 
     if(vf->type&VIDTYPE_VIU_422){
         if(interlace_mode == VIDTYPE_INTERLACE_TOP)
@@ -1774,6 +1774,7 @@ static void process_vf_change(vframe_t *vf, ge2d_context_t *context, config_para
     temp_vf.duration_pulldown = vf->duration_pulldown;
     temp_vf.pts = vf->pts;
     temp_vf.pts_us64 = vf->pts_us64;
+    temp_vf.flag = vf->flag;
     temp_vf.type = VIDTYPE_VIU_444 | VIDTYPE_VIU_SINGLE_PLANE | VIDTYPE_VIU_FIELD;
     temp_vf.canvas0Addr = temp_vf.canvas1Addr = ass_index;
     cur_angle = (ppmgr_device.videoangle + vf->orientation)%4;
diff --git a/include/linux/amlogic/amports/vframe.h b/include/linux/amlogic/amports/vframe.h
index e371ed1cfbdd..4dc66e2e0274 100755
--- a/include/linux/amlogic/amports/vframe.h
+++ b/include/linux/amlogic/amports/vframe.h
@@ -58,6 +58,8 @@
 #define DISP_RATIO_ASPECT_RATIO_BIT     8
 #define DISP_RATIO_ASPECT_RATIO_MAX     0x3ff
 
+#define VFRAME_FLAG_NO_DISCONTINUE      1
+
 typedef enum pixel_aspect_ratio_e {
         PIXEL_ASPECT_RATIO_1_1,
         PIXEL_ASPECT_RATIO_8_9,
-- 
2.19.0

